## Part B: Training and career development {.page_break_before}

### Publication Strategy

- The paper covering the work in
  [Chapter 1](./20.partA.md#chapter-1-variant-discovery-in-genome-graphs) is currently
  in preparation - titled *Nucleotide-resolution bacterial pan-genomics with reference
  graphs*. This paper covers the entirety of the `pandora` method. As the bulk of this
  method was produced by a previous PhD student in the lab, Rachel Colquhoun, I will be
  the second author. The work I will have contributed to this paper include the addition
  of the *de novo* variant discovery, and a large amount of the evaluation. We aim to
  submit the paper in October 2020.
- The work that will constitute
  [Chapter 2](./20.partA.md#chapter-2-applications-to-m-tuberculosis-nanopore-variant-calling)
  and
  [Chapter 3](./20.partA.md#chapter-3-applications-to-improving-m-tuberculosis-drug-resistance-prediction)
  will also be contained in a paper, of which I will be the first author. We aim to have
  this work completed and a manuscript submitted by the end of November 2020.
- The work in
  [Chapter 4](./20.partA.md#chapter-4-construction-of-a-m-tuberculosis-reference-pan-genome)
  is a bit harder to put an approximate date on, but we would hope to have it completed
  sometime in the third quarter of 2021, with myself being the first author.

### List of publications, papers in press, preprints, manuscripts submitted/in preparation to date

In addition to the major "thesis publications" listed [above](#publication-strategy), I
am also involved in the following publications:

- The recent publication of `mykrobe` [@doi:10/ggd835], in which I assisted with the
  improvement of the genotyping model for Nanopore data.
- A preprint [@doi:10.1101/2020.02.06.936302] from a student-led project investigating
  the use of Nanopore sequencing for freshwater monitoring. I did a large part of the
  data analysis for this work.
- Supplementary to the work in
  [Chapter 2](20.partA.md#chapter-2-applications-to-m-tuberculosis-nanopore-variant-calling)
  we also aim to submit a paper that will be a "crowd-sourcing" call for training
  species-specific Nanopore basecalling models. I used the data we have good PacBio
  assemblies for to try and train a Mtb-specific basecalling model, as it has been
  previously shown (in *K. pneumoniae*) [@doi:10.1186/s13059-019-1727-y] that this can
  improve read and consensus accuracy. This attempt did not yield a more accurate model
  than the default, but we think it is important to make this result available and
  challenge others to do the same for their species-of-interest.
- I am third-author on a new manuscript for the workflow management system `snakemake`
  [@doi:10.1093/bioinformatics/bts480]. The manuscript is in the process of being
  submitted to *Nature Communications*. I contributed to this work by co-developing a
  tool called `snakefmt` (<https://github.com/snakemake/snakefmt>) that is used to
  provide automatic formatting of `snakemake` code - improving the readability of
  workflows. Additionally, I also developed and maintain the `snakemake`
  ["profile"][lsf] that configures job submission and status-checking for the LSF
  cluster system.
- I plan to submit an ["Application Note"][app-note] to *Bioinformatics* for `rasusa`
  [@doi:10/d9rz]. This is a tool I wrote in the Rust programming language to randomly
  subsample sequencing reads to a specified coverage.

### Work plan and timeline for thesis submission

> Provide a work plan to be completed prior to thesis writing and a timeline for
> submission.

### List of scientific courses and conferences attended to date and planned for next year

> Please list all events in reverse chronological order.

### List of additional training, teaching and other relevant activities to date

> Please list all training and other activities in reverse chronological order.

### Career development plan

#### Please describe your current long-term career aims (i.e. 3-5 years after PhD).

#### Please comment on the types of position you would like to apply to for after the PhD and your expected application timeline?

> If applying for postdoc positions, which fields/fellowships are you considering? If
> you are applying for non-academic careers, what type of role? Do you have target
> companies or organisations in mind?

#### What do you see as your strengths (2-3 skills)?

#### What do you see as your areas for improvement1 (2-3 areas)?

#### What are your career development priorities until the end of your contract?

> Please take into account both the scientific and non-scientific skills you will need
> to work on a successful and timely completion of the PhD. List the skills you still
> need to acquire to achieve your longer-term career aims.

#### What actions will you take to develop these skills?

> This may include training courses, opportunities to practice and develop these skills,
> or seeking feedback/guidance. N.B. fellows should take 1-2 non-scientific trainings or
> career development workshops per year, and at least one international conference
> during their PhD.

[app-note]: https://academic.oup.com/bioinformatics/pages/instructions_for_authors#Types%20of%20Manuscript
[lsf]: https://github.com/Snakemake-Profiles/lsf
