<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-UK" xml:lang="en-UK">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Michael B. Hall" />
  <meta name="dcterms.date" content="2020-09-23" />
  <meta name="keywords" content="phd, thesis-plan, genome-graphs, bacterial-genomics, nanopore, tuberculosis, variant-calling, drug-resistance" />
  <title>Third-year progress report for thesis advisory committee</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="Third-year progress report for thesis advisory committee" />
  <meta name="citation_title" content="Third-year progress report for thesis advisory committee" />
  <meta property="og:title" content="Third-year progress report for thesis advisory committee" />
  <meta property="twitter:title" content="Third-year progress report for thesis advisory committee" />
  <meta name="dc.date" content="2020-09-23" />
  <meta name="citation_publication_date" content="2020-09-23" />
  <meta name="dc.language" content="en-UK" />
  <meta name="citation_language" content="en-UK" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="Michael B. Hall" />
  <meta name="citation_author_institution" content="EMBL-EBI" />
  <meta name="citation_author_institution" content="University of Cambridge" />
  <meta name="citation_author_orcid" content="0000-0003-3683-6208" />
  <meta name="twitter:creator" content="@mbhall88" />
  <link rel="canonical" href="https://mbhall88.github.io/TAC3_Report/" />
  <meta property="og:url" content="https://mbhall88.github.io/TAC3_Report/" />
  <meta property="twitter:url" content="https://mbhall88.github.io/TAC3_Report/" />
  <meta name="citation_fulltext_html_url" content="https://mbhall88.github.io/TAC3_Report/" />
  <meta name="citation_pdf_url" content="https://mbhall88.github.io/TAC3_Report/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://mbhall88.github.io/TAC3_Report/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://mbhall88.github.io/TAC3_Report/v/81491c984eb4d0d3fe4b6de5801e715e43f585a4/" />
  <meta name="manubot_html_url_versioned" content="https://mbhall88.github.io/TAC3_Report/v/81491c984eb4d0d3fe4b6de5801e715e43f585a4/" />
  <meta name="manubot_pdf_url_versioned" content="https://mbhall88.github.io/TAC3_Report/v/81491c984eb4d0d3fe4b6de5801e715e43f585a4/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Third-year progress report for thesis advisory committee</h1>
</header>
<p><small><em>
This manuscript
(<a href="https://mbhall88.github.io/TAC3_Report/v/81491c984eb4d0d3fe4b6de5801e715e43f585a4/">permalink</a>)
was automatically generated
from <a href="https://github.com/mbhall88/TAC3_Report/tree/81491c984eb4d0d3fe4b6de5801e715e43f585a4">mbhall88/TAC3_Report@81491c9</a>
on September 23, 2020.
</em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><strong>Michael B. Hall</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-3683-6208">0000-0003-3683-6208</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/mbhall88">mbhall88</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/mbhall88">mbhall88</a><br>
<small>
EMBL-EBI; University of Cambridge
· Funded by EMBL International PhD Programme (EIPP)
</small></li>
</ul>
<h2 class="page_break_before" id="thesis-advisory-committee">Thesis Advisory Committee</h2>
<ul>
<li>Zamin Iqbal (Supervisor) - EMBL-EBI</li>
<li>John Marioni (Chair) - EMBL-EBI</li>
<li>Georg Zeller - EMBL Heidelberg</li>
<li>Estée Török - University of Cambridge</li>
</ul>
<p><strong>Starting Date</strong>: 12/10/2017<br />
<strong>Qualifying Assessment Date</strong>: 06/07/2018<br />
<strong>Second TAC Meeting</strong>: 15/10/2019<br />
<strong>Third TAC Meeting</strong>: 13/10/2020</p>
<h2 class="page_break_before" id="part-a-progress-report">Part A: Progress Report</h2>
<h3 id="examining-bacterial-variation-with-genome-graphs">Examining bacterial variation with genome graphs</h3>
<h3 id="executive-summary">Executive summary</h3>
<p>Genomics is now ubiquitous in clinical and public health microbiology, at least in the
developed world. However, many significant challenges remain.</p>
<ul>
<li>Bacterial genomes harbour huge amounts of diversity, even within a species, and
traditional reference-based approaches are problematic.</li>
<li>Much of the variation in bacteria is fundamentally inaccessible to short reads.</li>
<li>Long Nanopore reads are noisy, and SNP calling with this data is not properly
benchmarked or standardised.</li>
<li>Since <em>Mycobacterium tuberculosis</em> (Mtb) infects so many people, there is potential
for considerable impact for clinical applications.</li>
<li>There is also much to be gained from a high-quality pan-genome of Mtb as well as a
detailed map of its enigmatic <em>pe/ppe</em> gene repertoire.</li>
</ul>
<p>These motivations drive the following PhD thesis structure:</p>
<ol type="1">
<li>Develop algorithms and software for variant discovery using bacterial genome graphs,
building on work of a previous student in the lab (my first paper, second author).</li>
<li>Benchmark Nanopore versus Illumina SNP calling, showing our algorithms meet the needs
of clinical and public health users, validate, and publish (second paper).</li>
<li>Improve upon current whole-genome sequencing-based drug resistance prediction for Mtb
using genome graphs.</li>
<li>Curate a high-quality reference pan-genome for Mtb that includes a detailed map of
the <em>pe/ppe</em> genes.</li>
</ol>
<!--================================================================================-->
<h3 id="motivation-and-background">Motivation and Background</h3>
<p>In 2018, 1.4 million people died of tuberculosis (TB) globally, and over 10 million
people fell ill to the disease. There were also 484,000 new cases of
rifampicin-resistant TB reported, of which 78% were multi-drug resistant (MDR)
<span class="citation" data-cites="owU3wUrb">[<a href="#ref-owU3wUrb" role="doc-biblioref">1</a>]</span>. Drug-resistant TB is not a new problem; the first clinical trial
using Streptomycin to treat TB reported resistance and this outcome influenced the
creation of the four-drug first-line regimen used today. Standard of care requires
phenotypic testing of the infecting organism against these four drugs to ensure that
appropriate treatment is prescribed. However, <em>Mycobacterium tuberculosis</em> (Mtb), the
causative agent of TB, is a slow-growing organism and the gold-standard phenotypic tests
(“mix bug and drug”) take around two months to complete. Thus, the traditional clinical
diagnostic and treatment regimen is slow and expensive. Whole-genome sequencing offers a
faster solution; recently it was shown that equivalent results are achievable by
sequencing Mtb grown in “liquid culture” (also known as MGIT, Mycobacterial Growth
Indicator Tube) after two weeks of culture in contrast to the two-month traditional
(Lowenstein-Jensen) culture method <span class="citation" data-cites="Z3NG0hEI">[<a href="#ref-Z3NG0hEI" role="doc-biblioref">2</a>]</span>. A number of genes are
implicated in drug resistance and predicting resistance from sequencing data based on a
catalogue of resistance single-nucleotide polymorphisms (SNPs) and indels (insertions or
deletions) works with high-specificity <span class="citation" data-cites="d8RxFwPN 11Jcu87Ae">[<a href="#ref-d8RxFwPN" role="doc-biblioref">3</a>,<a href="#ref-11Jcu87Ae" role="doc-biblioref">4</a>]</span>. For the four
first-line drugs, a study by the CRyPTIC consortium is the first of its kind to
demonstrate that phenotyping is not required if genotype predicts susceptibility
<span class="citation" data-cites="4zIt0Vce">[<a href="#ref-4zIt0Vce" role="doc-biblioref">5</a>]</span>. However, as the genetic basis for drug resistance is not
entirely understood, there is still a sensitivity gap that differs drug-by-drug.</p>
<p>Public health requirements for TB diagnostics are <strong>resistance prediction, species
identification, and clustering</strong> of genomes. The clusters are intended to contain
potential transmission events, thus enabling more effective contact tracing. All of the
main requirements are currently successfully achieved with Illumina sequencing
technology. However, there are reasons to consider abandoning Illumina for Oxford
Nanopore Technology’s (ONT) sequencer. First, there is cost - Illumina has raised its
reagent prices considerably. Second, the burden of TB lies primarily in places where
there is neither capital nor infrastructure for purchasing or maintaining a large
machine. The third is speed: Votintseva <em>et al.</em> showed that it was feasible to go from
patient to result via a Nanopore sequencer in 12.5 hours <span class="citation" data-cites="9NLFJZN2">[<a href="#ref-9NLFJZN2" role="doc-biblioref">6</a>]</span>.
Since this publication, the yield and quality of Nanopore has risen. Therefore, a
diagnostic that delivers results while the patient is in the clinic is now imaginable.</p>
<p>As an aside, it is worth noting that in high prevalence areas, there is no interest in
transmission analysis, and therefore all of the information that they need (species and
resistance) could be attained via deep amplicon sequencing. We accept this but ignore it
and focus on a whole-genome solution nevertheless; both because of the better
phylogenetic resolution, and because we do not yet know the relevant genetic mutations
associated with new drugs.</p>
<p>The idea of using a single, linear reference genome to represent a population of
individuals is not ideal. Even more so in the context of a bacterium such as <em>Salmonella
enterica</em>, where two individuals can differ to such a degree that they only share 16% of
their genes <span class="citation" data-cites="5WDPm9pG IYW3CNeu">[<a href="#ref-5WDPm9pG" role="doc-biblioref">7</a>,<a href="#ref-IYW3CNeu" role="doc-biblioref">8</a>]</span>. This fluidity of genetic content leads to
the concept of a pan-genome; defined as the set of all genes observed in a species. Some
genes are more common than others within this pan-genome, leading to the distinction of
the ‘core’ and ‘accessory’ genome with the core being those shared across (most) all
individuals and accessory being everything else. The effect of using a single-reference
genome to describe variation within samples from the same species is best illustrated in
Figure <a href="#fig:refbias">1</a>. In this toy example, the maximum number of SNPs we can hope to
discover is only 56%. Clearly, comparing many samples requires better methods than
single-reference based ones.</p>
<div id="fig:refbias" class="fignos">
<figure>
<img src="images/refbias.png" style="height:4in" alt="" /><figcaption><span>Figure 1:</span> An illustration of how reference bias can impact variant-calling. Each vertical
column signifies an individual genome, with the coloured blocks representing genes.
Numbers label 50 segregating SNPs. The percentages at the bottom express the proportion
of SNPs which can be detected using each of the 6 ‘genomes’ as a reference by mapping
perfect reads from the remaining 5 to this reference.</figcaption>
</figure>
</div>
<p>One of the first questions that arise from a computational point-of-view is how to apply
current methods, which assume a single reference, to a pan-genome? An approach to
answering this question that has been gaining considerable traction in recent years is
that of genome graphs. This new paradigm uses a population reference graph (PRG) as its
equivalent of a single, linear, reference genome. A PRG is built to represent variation
seen within a population, conceding the fact that no single genome can accurately
represent an entire species. To construct such a PRG, one takes a multiple sequence
alignment (MSA) and collapses shared sequence and creates “bubbles”, or branch points,
where they do not (see Figure <a href="#fig:prg">2</a> for an illustration of this process). We define
a collection of PRGs as a pan-genome reference graph (PanRG), which we will
interchangeably refer to as a pan-genome. Thus far, genome graph methods have focused
mainly on eukaryotes. Given the rich diversity of genetic content in the prokaryote
world, it would seem genome graphs are more suited to utilisation there.</p>
<div id="fig:prg" class="fignos">
<figure>
<img src="images/prg.png" style="width:100.0%" alt="" /><figcaption><span>Figure 2:</span> An illustration of how a population reference graph (PRG) is constructed. Regions
(columns) of shared sequence are collapsed into a single node. Those that differ are
split into “bubbles”, or branching nodes.</figcaption>
</figure>
</div>
<p>Nanopore sequencing yields very long reads (up to 2.2Mbp <span class="citation" data-cites="QrKrCCsB">[<a href="#ref-QrKrCCsB" role="doc-biblioref">9</a>]</span>), the
mean/mode read identity tends to fall in the range of 87-94% (using ONT’s <code>guppy</code>
basecaller), while on a consensus level, it can achieve 99.94% with assembly polishing
<span class="citation" data-cites="lwFwJNS9">[<a href="#ref-lwFwJNS9" role="doc-biblioref">10</a>]</span>.<br />
Variant calling with Nanopore sequencing data has seen a somewhat slow development.
Currently, the main tools that have had reasonable testing done and are versatile enough
to detect both SNPs and (some) indels are <code>nanopolish</code> <span class="citation" data-cites="Oe1xiQra">[<a href="#ref-Oe1xiQra" role="doc-biblioref">11</a>]</span>, Clair
<span class="citation" data-cites="LurPKJ01">[<a href="#ref-LurPKJ01" role="doc-biblioref">12</a>]</span> and <code>medaka</code> <span class="citation" data-cites="VSoEWPFn">[<a href="#ref-VSoEWPFn" role="doc-biblioref">13</a>]</span>. A recent
benchmark showed that Nanopore variant calling provides reliable diagnostic information
for <em>Neisseria gonorrhoeae</em> <span class="citation" data-cites="pPxMDtt0">[<a href="#ref-pPxMDtt0" role="doc-biblioref">14</a>]</span>. However, to date, there has
been no extensive Nanopore variant calling benchmark done for Mtb. Given the potential
benefits of using genome graphs and long-read Nanopore sequencing for bacterial
genomics, it makes sense to try and blend the two.</p>
<p>Pandora is a method being developed in the group to genotype across the <em>entire</em>
pan-genome of a bacterial sample - not just the core. It does this by working with a
PRG, rather than a linear reference. The method is based on the following intuition
(similar to that behind the Li and Stephens model in population genetics): genomes
evolve by recombination and mutation, and thus we ought to be able to approximate a
<span class="math inline">\(N+1\)</span> genome as a mosaic of the first <span class="math inline">\(N\)</span> genomes. <code>pandora</code> maps Nanopore reads to a
graph encoding of a PRG, infers a mosaic, and provides genotypes at all variants in the
PanRG. Mapping is done using minimising k-mers <span class="citation" data-cites="19hwdhKN">[<a href="#ref-19hwdhKN" role="doc-biblioref">15</a>]</span> in a similar vein to
that done by <code>minimap</code><span class="citation" data-cites="1DgB6gqeS">[<a href="#ref-1DgB6gqeS" role="doc-biblioref">16</a>]</span>, and is therefore fast. By using Nanopore
sequencing data, it is also possible to infer gene order as, in general, a single read
will contain multiple genes, as opposed to Illumina sequencing where multiple reads are
required to span a single gene. Note <code>pandora</code> does not (prior to this work) include any
facility for discovering novel variation.</p>
<h4 id="summary">Summary</h4>
<p>This PhD seeks to develop new methods to enable Nanopore-based diagnostics and
epidemiology for Mtb and other bacteria. By using PRGs as a strong prior <span class="citation" data-cites="2FYPCKEv 7AwGVQXO">[<a href="#ref-2FYPCKEv" role="doc-biblioref">17</a>,<a href="#ref-7AwGVQXO" role="doc-biblioref">18</a>]</span>, we should be able to mitigate the Nanopore error biases and indel
issues. In the process, we aim to construct a high-quality reference pan-genome for Mtb
that we hope will open previously inaccessible parts of its genome for investigation.</p>
<!--================================================================================-->
<h3 id="chapter-1-variant-discovery-in-genome-graphs">Chapter 1: Variant discovery in genome graphs</h3>
<p>Variation in bacterial genomes can arise through a diverse range of processes. Mutations
can arise during replication and are inherited vertically, genetic material can be
transferred horizontally, and homologous recombination can lead to eukaryote-like gene
conversion <span class="citation" data-cites="h4yvs7gl">[<a href="#ref-h4yvs7gl" role="doc-biblioref">19</a>]</span>. This breadth of ways in which bacteria can acquire new and
varied genetic material results gives rise to the phenomenon of a pan-genome. Bacterial
species with an “open” pan-genome may have individuals in their population who can share
as little as 16% of their genes (<em>S. enterica</em>) <span class="citation" data-cites="5WDPm9pG">[<a href="#ref-5WDPm9pG" role="doc-biblioref">7</a>]</span>. Not all species’
pan-genomes are this open, but it does raise the question: what do we use as a
“reference” genome for such a species? One solution is to use the reference genome for
the specific strain of interest. This works fine when dealing with a single sample or
multiple samples of the same strain. However, when expanding to many samples from
varying strains, the reference is no longer representative. An alternative solution is
to focus solely on the core genome - the complement of genes found in all members of a
species (see Figure <a href="#fig:pangenome">3</a>). The issue with this approach is the loss of
information about variation in all of the non-core genes, which could be a large number
if the pan-genome is open. See Figure <a href="#fig:pangenome">3</a> for an illustration of this
reference-bias problem.</p>
<div id="fig:pangenome" class="fignos">
<figure>
<img src="images/pangenome.jpg" style="width:50.0%" alt="" /><figcaption><span>Figure 3:</span> Schematic representation of a pan-genome. Each oval represents the
genetic content of a single individual within a species. Reference <span class="citation" data-cites="5WDPm9pG">[<a href="#ref-5WDPm9pG" role="doc-biblioref">7</a>]</span>.</figcaption>
</figure>
</div>
<p>The message we are aiming to drive home at the beginning of this chapter is that
reference-based analysis fundamentally cannot access all of the variation in a bacterial
species.</p>
<h4 id="prior-work-mosaic-approximations-and-genotyping">Prior work: Mosaic approximations and genotyping</h4>
<p>Standard approaches to variant analysis are effectively a first-order approximation. In
such an approximation, samples are considered identical to the reference, one maps reads
to it, identifies clear SNPs via the read pileup, and then modifies the reference to get
an estimate of the sample’s genome. As mentioned in
<a href="#motivation-and-background">Motivation and Background</a>, <code>pandora</code> is a method developed
by a previous PhD student in the lab, Rachel Colquhoun. It works on the premise of
approximating a genome as a hierarchical mosaic. At a high-level, it represents a mosaic
of loci - usually genes and intergenic regions - while at the locus-level, it is a
mosaic of previously-seen genomes.</p>
<p><code>pandora</code> has two main workflows: <code>map</code> and <code>compare</code>. <code>map</code> aims to infer a consensus
sequence from a PRG for a single sample. The consensus sequence says two things about
the sample: the loci present in the sample, and the most likely path through the PRG for
each present locus. Optionally, <code>map</code> can run genotyping of the sample, providing a
Variant Call Format (VCF) file for the sample with respect to the consensus sequence (or
a user-provided sequence). <code>compare</code> aims to infer a consensus sequence that best
represents a given set of genomes and then genotype each sample with respect to this
inferred “average” genome. The advantage of this approach is that the consensus best
represents the specific samples provided. This allows for genotyping across the <em>entire</em>
pan-genome. If a gene is present in only 2 of 50 samples then genotyping information is
provided for those 2 samples and null for the other 48.</p>
<p>While <code>pandora</code>, before the work in this chapter, allows comparison of genomes to a
level of detail provided by no other tool, there is still a significant shortcoming: it
cannot discover novel variation. If a sample contains a variant not present in the PRG,
the best <code>pandora</code> can do is select the path that is closest to that variant. The work
in my first chapter outlines a method for removing this limitation within <code>pandora</code> and
provides an analysis of the gain in recall and precision by incorporating <em>de novo</em>
variant discovery into the <code>pandora</code> workflow.</p>
<h4 id="local-de-novo-variant-discovery-in-a-genome-graph">Local <em>de novo</em> variant discovery in a genome graph</h4>
<p>There are two significant difficulties in discovering <em>de novo</em> variants on a graph. The
first is finding regions within the graph that look like the reads mapping to them
contain variation we do not have in the PRG. Secondly, we need to generate new paths for
these regions and add them back into the PRG for consideration when remapping.</p>
<h5 id="finding-candidate-regions">Finding candidate regions</h5>
<p>We define a candidate region, <span class="math inline">\(r\)</span>, as an interval within a local graph where coverage on
the maximum likelihood path is less than a given threshold, <span class="math inline">\(c\)</span>, for more than <span class="math inline">\(l\)</span>
consecutive positions. For a given read that has a mapping to <span class="math inline">\(r\)</span>, we define <span class="math inline">\(s_r\)</span> to be
the subsequence of the read mapping to <span class="math inline">\(r\)</span>. We define the pileup <span class="math inline">\(P_r\)</span> as the set of all
<span class="math inline">\(s_r \in r\)</span>.</p>
<h5 id="enumerating-paths-through-candidate-regions">Enumerating paths through candidate regions</h5>
<p>For <span class="math inline">\(r \in R\)</span>, where <span class="math inline">\(R\)</span> is the set of all candidate regions, we construct a de Bruijn
graph <span class="math inline">\(G_r\)</span> from <span class="math inline">\(P_r\)</span>.<br />
<span class="math inline">\(A_L\)</span> and <span class="math inline">\(A_R\)</span> are defined as kmers to the left and right of <span class="math inline">\(r\)</span> in the local graph.
They are anchors to allow re-insertion of new sequences found by <em>de novo</em> discovery
into the local graph. If <span class="math inline">\(A_L \notin G_r \lor A_R \notin G_r\)</span> then we abandon <em>de novo</em>
discovery for <span class="math inline">\(r\)</span>.<br />
<span class="math inline">\(T_r\)</span> is the spanning tree obtained by performing depth-first search (DFS) on <span class="math inline">\(G_r\)</span> from
node <span class="math inline">\(A_L\)</span>. <span class="math inline">\(p_r\)</span> is defined as a path, from the root node <span class="math inline">\(A_L\)</span> of <span class="math inline">\(T_r\)</span> and ending at
node <span class="math inline">\(A_R\)</span>, which fulfills the following two conditions:</p>
<ul>
<li><span class="math inline">\(p_r\)</span> is shorter than the maximum allowed path length.</li>
<li>No more than <span class="math inline">\(k\)</span> nodes along <span class="math inline">\(p_r\)</span> have coverage <span class="math inline">\(&lt; (n_r \times 0.1) \times e_r\)</span>,
where <span class="math inline">\(n_r\)</span> is the a counter, described below, and <span class="math inline">\(e_r\)</span> is the expected k-mer
coverage for <span class="math inline">\(r\)</span>.</li>
</ul>
<p><span class="math inline">\(V_r\)</span> is the set of all <span class="math inline">\(p_r\)</span>. If <span class="math inline">\(|V_r|\)</span> is greater than a predefined threshold, <span class="math inline">\(n_r\)</span>
is incremented by 1 and <span class="math inline">\(V_r\)</span> is repopulated. If <span class="math inline">\(n_r \times 0.1 = 1.0\)</span> then <em>de novo</em>
discovery is abandoned for <span class="math inline">\(r\)</span>.</p>
<h5 id="pruning-the-path-space-in-a-candidate-region">Pruning the path-space in a candidate region</h5>
<p>As <code>pandora</code> operates on both accurate and error-prone sequencing reads, the number of
valid paths in <span class="math inline">\(G_r\)</span> can be very large. In testing, this results in run-times beyond
seven days (the longest any attempt was allowed to run). The increased run-time is due
to cycles that can occur in <span class="math inline">\(G_r\)</span> and exploring paths that will never reach our required
end anchor <span class="math inline">\(A_R\)</span>. In order to reduce the path-space within <span class="math inline">\(G_r\)</span> we prune paths based on
multiple criteria. Critically, this pruning happens at each step of the graph walk
(path-building).<br />
In addition to <span class="math inline">\(T_r\)</span>, obtained by performing DFS on <span class="math inline">\(G_r\)</span>, we produce a distance map
<span class="math inline">\(D_r\)</span> that results from running reversed breadth-first search (BFS) on <span class="math inline">\(G_r\)</span>, beginning
from node <span class="math inline">\(A_R\)</span>. We say reversed BFS as we explore the predecessors of each node, rather
than the successors. <span class="math inline">\(D_r\)</span> is implemented as a binary search tree where each node in the
tree represents a k-mer in <span class="math inline">\(G_r\)</span> that is reachable from <span class="math inline">\(A_R\)</span> via reversed BFS. Each
node additionally has an integer attached to it that describes the shortest path from
that node to <span class="math inline">\(A_R\)</span>.<br />
We can use <span class="math inline">\(D_r\)</span> to prune the path-space by requiring, for each node, <span class="math inline">\(n \in p_r\)</span>: is <span class="math inline">\(n \in D_r\)</span> <strong>and</strong> can <span class="math inline">\(A_R\)</span> be reached from <span class="math inline">\(n\)</span> in a minimum of <span class="math inline">\(i\)</span> nodes, where <span class="math inline">\(i\)</span> is
defined as the maximum allowed path length minus the number of nodes walked to reach
<span class="math inline">\(n\)</span>. If one of these conditions is not met, we abandon <span class="math inline">\(p_r\)</span>. The advantage of this
pruning process is that we never explore paths that will not reach our required end
point and when caught in a cycle, we will abandon the path once we have made too many
iterations around the cycle.</p>
<p>I programmed the above methods in C++ and added them into the code base for <code>pandora</code>.
They constitute 1325 lines of source code and 3486 lines of test code. I had help with
the implementation of multi-threading the <em>de novo</em> component and the BFS pruning from
Leandro Ishi. The overall workflow for <code>pandora</code> with <em>de novo</em> variant discovery
enabled is illustrated in Figure <a href="#fig:pandora">4</a>.</p>
<div id="fig:pandora" class="fignos">
<figure>
<img src="images/pandora.png" style="width:80.0%" alt="" /><figcaption><span>Figure 4:</span> Schematic of the <code>pandora</code> method. Note Steps 4-6 are those developed and added
in the scope of chapter 1. 1) A Population Reference Graph (PRG) built from known
variants or previous samples. 2) Sequencing reads to try and map on to the PRG. 3) Reads
mapped on to the PRG (where possible) using minimising k-mers. Coloured segments show
paths where reads map. 4) Local assembly of the reads using a de Bruijn graph due to
lower than expected coverage on a section of the PRG (many unmapped reads). A new path
is discovered (orange). 5) The new path is added into the original PRG. 6) Reads are
remapped to the new PRG with many reads mapping to the newly included path (orange).</figcaption>
</figure>
</div>
<h4 id="evaluation">Evaluation</h4>
<h5 id="simulated-data">Simulated data</h5>
<p>The first step in evaluating the effect of adding <em>de novo</em> variant calling to <code>pandora</code>
is with a simulated dataset. What we aim to show here is that the addition of <em>de novo</em>
discovery allows <code>pandora</code> to improve its probability of variant detection (recall). The
first step was randomly selecting 100 gene MSAs from a pool of 29702 obtained for
<em>Escherichia coli</em> from the panX database <span class="citation" data-cites="IYW3CNeu">[<a href="#ref-IYW3CNeu" role="doc-biblioref">8</a>]</span>. Next, a PRG is constructed
for each gene MSA, and a random path through each is selected using <code>pandora</code>. Each
PRG’s random path is then concatenated together to form a single “genome” sequence. We
subsequently add SNPs to the simulated genome at a specified rate of SNPs per-gene using
<code>snp-mutator</code> <span class="citation" data-cites="QdpEzyJH">[<a href="#ref-QdpEzyJH" role="doc-biblioref">20</a>]</span>. We then simulated Nanopore reads from this mutated genome
using <code>nanosim-h</code> <span class="citation" data-cites="Fp8BI1Lc 7m2CKnnh">[<a href="#ref-Fp8BI1Lc" role="doc-biblioref">21</a>,<a href="#ref-7m2CKnnh" role="doc-biblioref">22</a>]</span>. <code>pandora map</code> is then run, using the
original PRG as input, along with the reads simulated from the mutated genome. With this
approach, we know the exact SNPs we hope to find. After running the <code>map</code> routine, we
are left with a collection of candidate paths produced by the <em>de novo</em> component. We
then added these back into the PRG and iterated the <code>map</code> routine again, this time
without <em>de novo</em> discovery enabled. Re-adding variants into the graph has the effect of
making all of the <em>de novo</em> variants visible to genotyping. The result is a VCF file
that (hopefully) contains the variants we introduced in the beginning. To compare the
performance of <code>pandora</code> before and after <em>de novo</em> is added, we can compare the VCF
produced from the first round of <code>map</code> with the last VCF.</p>
<p>To avoid error-prone conversion of linear coordinates into graph coordinates the
evaluation of whether the variants called by <code>pandora</code> are correct was undertaken in a
slightly more convoluted manner. We define a probe-set <span class="math inline">\(P\)</span> as a collection of probes,
<span class="math inline">\(p\)</span>, where <span class="math inline">\(p\)</span> represents an entry, <span class="math inline">\(e\)</span>, in a VCF file, <span class="math inline">\(V\)</span>. For each <span class="math inline">\(e \in V\)</span>, <span class="math inline">\(p\)</span> is
constructed by the concatenation of <span class="math inline">\(l_w\)</span>, <span class="math inline">\(e_c\)</span>, and <span class="math inline">\(r_w\)</span> (in that order), where <span class="math inline">\(e_c\)</span>
is the called variant of <span class="math inline">\(e\)</span>, and <span class="math inline">\(l_w\)</span> and <span class="math inline">\(r_w\)</span> are the sequences, of maximum length
<span class="math inline">\(w\)</span>, in the VCF reference to the left and right, respectively, of <span class="math inline">\(e_c\)</span>.</p>
<p>A truth probe-set, <span class="math inline">\(P_t\)</span>, was constructed from the VCF of variants added to the
simulated genome and a query probe-set, <span class="math inline">\(P_q\)</span>, from the variants called by <code>pandora</code>. We
then mapped all probes from <span class="math inline">\(P_t\)</span> to <span class="math inline">\(P_q\)</span> using <code>bwa mem</code> <span class="citation" data-cites="gmH6YDca">[<a href="#ref-gmH6YDca" role="doc-biblioref">23</a>]</span>. We then
classify each mapped probe as a false positive or true positive and calculate precision
and recall for the pre-<em>de novo</em> and post-<em>de novo</em> VCF files from <code>pandora</code>. As
expected, Figure <a href="#fig:simulation_roc">5</a> shows that without <em>de novo</em> variant discovery,
we are unable to find almost all introduced variants. In future work, we plan to add
indels to the simulations.</p>
<div id="fig:simulation_roc" class="fignos">
<figure>
<img src="images/simulation_roc_denovo_and_no_denovo.png" style="width:90.0%" alt="" /><figcaption><span>Figure 5:</span> Precision-recall curve for <code>pandora</code> simulations with (circles) and without (crosses) <em>de novo</em> variant
discovery. This plot is measuring the proportion of SNPs we introduced into our
simulated genome we correctly found (recall) on the X-axis. On the Y-axis is the
proportion of calls we made which were correct (precision). Colours represent different
coverage depth. Each point represents a different genotype confidence
threshold.</figcaption>
</figure>
</div>
<h5 id="empirical-data---multi-sample-comparison">Empirical data - multi-sample comparison</h5>
<p>This section will be the major focus for the evaluation of both <em>de novo</em> <em>and</em>
<code>pandora</code> and will constitute the results section of the <code>pandora</code> paper (near
completion). We aim to show, with the <em>compare</em> routine, the power of using a PRG
instead of a single reference along with <code>pandora</code>’s performance compared to
<code>nanopolish</code> and <code>medaka</code> (Nanopore) and <code>snippy</code> (Illumina)
<span class="citation" data-cites="6S74PtQr">[<a href="#ref-6S74PtQr" role="doc-biblioref">24</a>]</span>. We compare 20 samples, from across the <em>E. coli</em>
phylogeny. As other variant callers do not use a PRG we carefully selected 24
single-reference genomes representing the diversity of the <em>E. coli</em> phylogenetic tree
as best as possible. Each variant caller was run once for each reference genome with the
aim being to show that the more samples in the analysis, the lower the number of
variants the caller can find.</p>
<p>The evaluation will again follow a similar method to the
<a href="#simulated-data">Simulated data</a>. The truth for this analysis, however, is a bit more
complicated. In the process of developing an evaluation framework, we produced a python
package called <code>varifier</code> (<a href="https://github.com/iqbal-lab-org/varifier" class="uri">https://github.com/iqbal-lab-org/varifier</a>). Briefly, for
precision, <code>varifier</code> creates ‘probes’ for each variant in the VCF, using the genome the
variants we called with-respect-to. It then maps these probes to the truth genome for
the sample and determines the distance between the variant component of the probe and
the part of the truth genome it maps to. For recall evaluation, <code>varifier</code> collects all
differences in the pairwise alignment between the truth and VCF-reference genomes.
Probes are created for these differences (based on the truth genome) and they are mapped
to an augmented version of the VCF-reference genome, which has had the variants applied
to it. The mappings are then evaluated in the same way as for precision.<br />
As <code>pandora compare</code> calls variants for each sample with respect to an inferred best
approximation sequence for all samples, creating the set of truth variants is slightly
different. We perform a pairwise alignment for all pairs of samples and collect all the
differences from this alignment. We then deduplicate this panel to ensure variants are
not “double counted”, meaning core genome variants would have an unbalanced effect on
the overall precision and recall. We then follow the same probe-mapping approach from
<code>varifier</code> with these truth variants.</p>
<!--https://github.com/iqbal-lab/pandora1_paper/issues/209-->
<p>We filtered <code>pandora</code> variants based on the following criteria:</p>
<ul>
<li>Depth less than 15x</li>
<li>Less than 20% of reads are on one strand</li>
<li>50% or more of kmers on the allele have zero coverage</li>
</ul>
<p>After we have run this analysis, we will also most likely need to investigate applying
filters to the VCF file to remove any biases relating to problems like strand bias or
coverage.</p>
<p>Whilst I wrote nearly all of the code and associated tests for evaluating the recall for
this analysis, a lot of it has since been refactored by Leandro Ishi and by Martin Hunt
in the form of <code>varifier</code>. Due to the novelty of the method we are proposing, this has
taken quite a lot of time and thought. I wrote 1250 lines of code to evaluate the
methods, along with 3500 lines of test code to ensure there are no bugs in our
evaluation. Additionally, I built the original <code>snakemake</code> <span class="citation" data-cites="NcYZqBux">[<a href="#ref-NcYZqBux" role="doc-biblioref">25</a>]</span> pipeline of
approximately 3500 lines of codes to orchestrate the entire evaluation and simulations
(However, much of this has been rewritten by Leandro Ishi).</p>
<!--TODO add ROC and associated text-->
<!--TODO do we still want to do this?-->
<!--\subsubsection{CPU and Memory Performance} In this section, we will investigate the-->
<!--`pandora` CPU and memory performance impact of adding *de novo* variant discovery. Both-->
<!--in single- and multi-threaded modes.-->
<h4 id="outstanding-work">Outstanding work</h4>
<p>The major work still outstanding for this project is the direct integration of <em>de novo</em>
candidates back into the PRG. The current procedure requires a fairly laborious,
multi-step process for adding <em>de novo</em> candidates into the PRG, requiring the user to
run a separate pipeline from <code>pandora</code>. Ultimately this will need to be handled all
within the <code>pandora</code> program with no intervention from the user.<br />
To measure whether the reference bias effect in Figure <a href="#fig:refbias">1</a> is real and
significant, we plan to measure it in three species: <em>E. coli</em>, <em>Klebsiella pneumoniae</em>,
and <em>Streptococcus pneumoniae</em>. For each of the species, we will take a set of
high-quality reference genomes, and perform the pair-wise alignment-based evaluation
from the <code>pandora</code> <a href="#empirical-data---multi-sample-comparison">multi-sample comparison</a>.
This work will aid the message in the introduction to my first chapter. Much of this
work is being completed by Leandro Ishi, but it borrows heavily on some of the
evaluation code I have written for the multi-sample comparison.<br />
Lastly, there is an ongoing refinement of the <em>de novo</em> variant discovery process from
the work in Chapters 2 and 3. The analysis in these chapters lean heavily on <code>pandora</code>,
but for Mtb, which has a very different pan-genome to that of <em>E. coli</em> - which was used
for most of the development of <code>pandora</code>’s methods.</p>
<!--================================================================================-->
<!--TODO improve title-->
<h3 id="chapter-2-using-genome-graphs-and-nanopore-sequencing-to-define-transmission-clusters-for-m.-tuberculosis">Chapter 2: Using genome graphs and nanopore sequencing to define transmission clusters for <em>M. tuberculosis</em></h3>
<p>Public health applications for genome sequencing of Mtb generally focus on three
use-cases: species identification, prediction of drug resistance, and clustering of
samples for epidemiological purposes. In this chapter, we plan to focus on how the
methods developed in <code>pandora</code> can be used to improve clustering of samples - generally
referred to as “transmission clusters” - while, in the next chapter, we will address the
drug resistance prediction component. The intention is to be able to use Nanopore data
for public health. Therefore, this chapter will focus on a head-to-head comparison of
Nanopore and Illumina sequencing technologies for classifying transmission clusters Mtb.
What we would like to show with the work in this chapter is that, contrary to current
dogma, Nanopore sequencing technology has advanced to the point where it can be applied
to this use-case to a standard acceptable by public health authorities. The work that
will constitute this chapter (and the next) is a collaboration with researchers in
Oxford, Birmingham, Madagascar, and South Africa, but I will be performing all of the
analysis.</p>
<h4 id="data">Data</h4>
<p>As the work in this chapter (and the next) involves a direct comparison of two
sequencing technologies - Illumina and Nanopore - the DNA sequenced by both must be
identical so that we can be certain the technology is the only source of differences, if
any. Each sample was sequenced on both platforms from the same isolate and DNA
extraction. In total, we received 118 samples from Madagascar, 83 from South Africa, and
46 from the National Tuberculosis Reference Lab in Birmingham; giving us a total of 247
samples.<br />
As these samples are not reference isolates, we need to be able to compare both Illumina
and Nanopore to a truth. To establish how each platform compares and differs from the
truth, we have additionally sequenced 35 of the Malagasy isolates with PacBio and will
use the high-quality assemblies for these samples as a baseline comparison for samples
without a truth.</p>
<h4 id="quality-control">Quality Control</h4>
<p>The purpose of quality control (QC) is to ensure all samples used in later analysis are
of the highest quality. By highest quality we mean all samples have perfectly matched
Illumina and Nanopore data, sufficient coverage on both sequencing technologies, no
contamination, and no evidence of a mixed Mtb population.</p>
<p>The first step in QC was to exclude samples where the Nanopore and Illumina data were
not perfectly matched. There were some instances where isolates had to be resequenced as
the Nanopore data had been accidently deleted from the sequencing laboratory’s computer.
Additionally, some samples did not have any matched Nanopore or Illumina sequencing. In
total, we excluded 40 samples at this stage.</p>
<p>The remaining 207 samples were processed through a quality control pipeline
(<a href="https://github.com/mbhall88/head_to_head_pipeline/tree/master/data/QC" class="uri">https://github.com/mbhall88/head_to_head_pipeline/tree/master/data/QC</a>). The first
step in QC is decontamination of sequencing reads. We used the decontamination database
from <code>clockwork</code> (<a href="https://github.com/iqbal-lab-org/clockwork" class="uri">https://github.com/iqbal-lab-org/clockwork</a>), which contains a wide
range of organisms, including viral, human, Mtb, non-tuberculosis Mycobacterium (NTM),
and nasopharyngeal-associated bacterial genomes. Each genome has associated metadata
indicating if it is contamination or not. Reads were mapped to the database using <code>bwa mem</code> <span class="citation" data-cites="gmH6YDca">[<a href="#ref-gmH6YDca" role="doc-biblioref">23</a>]</span> (Illumina) and <code>minimap2</code> (Nanopore) <span class="citation" data-cites="172cJaw4Q">[<a href="#ref-172cJaw4Q" role="doc-biblioref">26</a>]</span>. The
resulting alignment was used to quantify the proportion of reads considered
contamination, unmapped, and wanted. A read is considered wanted if it has any mapping
to a non-contamination genome in the database and is output to a final decontaminated
fastq file. All other mapped reads are considered contamination.</p>
<p>All decontaminated fastq files were subsampled to a depth of 60x (Illumina) and 150x
(Nanopore) using <code>rasusa</code> <span class="citation" data-cites="4U7AuMzS">[<a href="#ref-4U7AuMzS" role="doc-biblioref">27</a>]</span>. The reason for subsampling is to limit
unnecessarily large read sets that can drastically slow down later steps in the analysis
process without significant advantage.</p>
<p>The last step in the QC pipeline is to assign lineages for each sample. A panel of
lineage-defining SNPs <span class="citation" data-cites="c17o0GrE 6J5Bt0f7 1CS78SxuA">[<a href="#ref-c17o0GrE" role="doc-biblioref">28</a>,<a href="#ref-6J5Bt0f7" role="doc-biblioref">29</a>,<a href="#ref-1CS78SxuA" role="doc-biblioref">30</a>]</span> was used in
conjunction with a sample’s Illumina VCF from the
<a href="#baseline-variant-analysis">Baseline variant analysis</a> for the lineage assignment. At
each lineage-defining position in the sample’s VCF we determine if the called allele is
the same as the panel allele. If it is, we add the full lineage that allele defines
(e.g. 4.1.1) to a list of called lineages. For this analysis, if more than one
heterozygous call was made at lineage-defining positions, we abandon lineage assignment
for that sample. After classifying all of a sample’s lineage-defining positions we then
produce a lineage assignment based on the list of called lineages. The most recent
common ancestor of all the called lineages is used as the lineage assignment. For
example, if the called lineages were <code>[4, 4.2.3, 4.2.5]</code> the lineage assignment would be
4.2. If there is more than one called lineage from a different major lineage group, a
mixed lineage assignment is given. For example <code>[4, 4.2.3, 4.2.5, 3.2]</code> would still be
called lineage 4.2, however, <code>[4, 4.2.3, 4.2.5, 3.2, 3.1]</code> would be called mixed.</p>
<p>In the end, we chose to exclude samples from further analysis if they met any of the
following criteria:</p>
<ul>
<li>Illumina coverage below 20x</li>
<li>Nanopore coverage below 30x</li>
<li>Evidence of mixed infection - i.e. mixed lineage classification</li>
<li>Unknown lineage assignment - no valid SNPs at lineage-defining sites</li>
</ul>
<p>This filtering criteria led to a further 57 samples being excluded; leaving us with a
total of 150 samples to use for the remainder of this work.</p>
<p>In addition to the QC of the Illumina/Nanopore data, we sadly had to exclude 26/35
PacBio sequencing datasets due to mismatched Illumina/Nanopore data or PacBio coverage
lower than 20x.</p>
<h4 id="genetic-clustering-of-samples">Genetic clustering of samples</h4>
<p>Although there is scientific interest in the question of identifying transmission chains
from genetic data, all the actionable public health information exists in the
identification of transmission clusters <span class="citation" data-cites="tGyXVFvJ 64Rp8zLg">[<a href="#ref-tGyXVFvJ" role="doc-biblioref">31</a>,<a href="#ref-64Rp8zLg" role="doc-biblioref">32</a>]</span>.</p>
<p>The first step towards clustering a set of genomes is determining a distance matrix.
Typically, this is done either by feeding aligned genomes into a phylogenetic
tree-building tool, or more coarsely, by merely counting SNP differences and clustering
based on these <span class="citation" data-cites="18CTtAeF9 tGyXVFvJ 64Rp8zLg">[<a href="#ref-tGyXVFvJ" role="doc-biblioref">31</a>,<a href="#ref-64Rp8zLg" role="doc-biblioref">32</a>,<a href="#ref-18CTtAeF9" role="doc-biblioref">33</a>]</span>. For the majority of
bacteria, there is a necessary step of identifying recombination tracts - which will
contain a high density of SNPs - and removing them. Removal of these SNPs is necessary
as they will have arrived at a different rate to the putative molecular clock and will
artefactually extend branch lengths on the phylogenetic tree
<span class="citation" data-cites="tGyXVFvJ 64Rp8zLg">[<a href="#ref-tGyXVFvJ" role="doc-biblioref">31</a>,<a href="#ref-64Rp8zLg" role="doc-biblioref">32</a>]</span>. In the case of Mtb, however, there is virtually no
recombination <span class="citation" data-cites="dRRsEE8Z hpSc80yh 1H760aM5J">[<a href="#ref-dRRsEE8Z" role="doc-biblioref">34</a>,<a href="#ref-hpSc80yh" role="doc-biblioref">35</a>,<a href="#ref-1H760aM5J" role="doc-biblioref">36</a>]</span>, so this step in not
required.</p>
<p>There are three main issues we need to address or keep in mind for the clustering
component of this chapter:</p>
<ul>
<li>When running a variant caller on a single sample, typically the tool only makes a
non-reference call when it finds a definite difference from the reference. Therefore,
it is impossible to tell the difference between a reference-matching site and one in
which there is genotyping uncertainty. One solution is to make a de-duplicated list of
variants found in all samples and genotype the samples at those positions. Solving
this reference-bias issue is the strength of the <code>compare</code> routine within <code>pandora</code>.
It tackles the problem in this manner while also handling issues of overlapping
variants in the list of all-sample-variants.</li>
<li>What distance measure do we use? Do we exclude positions where any sample has missing
data? Missing data of this type could be due to low coverage, but it could also be
because a section of the genome is not present. We will extensively cover this issue
for <em>E. coli</em> in <a href="#chapter-1-variant-discovery-in-genome-graphs">Chapter 1</a>. The
pan-genome of Mtb is small, but it does exist nonetheless <span class="citation" data-cites="Coyn9XyF 1H760aM5J">[<a href="#ref-1H760aM5J" role="doc-biblioref">36</a>,<a href="#ref-Coyn9XyF" role="doc-biblioref">37</a>]</span>. There is also a secondary issue that if certain regions have
clustered genetic variation in some lineages of Mtb then reads will not map well to
the reference. Both of the above issues are handled by <code>pandora</code> due to its use of a
PRG, but we still need to make a choice about missing data.</li>
<li>How do we handle sites where there is evidence of heterozygosity - i.e. a mixed
sample?</li>
</ul>
<p>For this chapter, we define genetic distance to be the sum of genetic discordances,
where missing data and heterozygosity do not cause discordance and study the clustering
this definition generates.</p>
<h4 id="multi-sample-comparison">Multi-sample comparison</h4>
<p>Multi-sample comparison suffers from two main challenges. First, large chunks of DNA may
be present or absent across samples - this is the pan-genome effect. This effect causes
significant issues with single-reference approaches, as outlined previously, but
<code>pandora</code> was developed to address this. Second, when comparing a set of samples, the
choice of reference affects how one describes variants. <code>pandora</code>, by design, chooses a
reference for each loci PRG based on the current dataset, intending to maximise the
succinctness of variant descriptions (see Figure <a href="#fig:refbias">1</a>). For example, we want
to see SNPs as what they are - single-base variants - not as a nested region (as in
Figure <a href="#fig:refbias">1</a>).</p>
<p>As mentioned, we handle these cases implicitly in <code>pandora</code>, however, the aim of this
chapter and the next, in addition to comparing across technologies, will also be to
compare methods for gaining drug resistance and epidemiological clustering information.
To be able to compare with other methods though, we will need to be able to compare
variants called with respect to a single-reference by other tools, with those from
<code>pandora</code>.</p>
<h4 id="baseline-variant-analysis">Baseline variant analysis</h4>
<p>The truth set of variants for the Illumina data in this chapter come from running the
Public Health England pipeline, COMPASS <span class="citation" data-cites="Lvhz1sqs">[<a href="#ref-Lvhz1sqs" role="doc-biblioref">38</a>]</span>. This pipeline will act as a
guide to inform us about whether the results from the Nanopore data are comparable with
those being used in real public health settings. COMPASS effectively uses <code>samtools</code>
<span class="citation" data-cites="IPc8cy7s">[<a href="#ref-IPc8cy7s" role="doc-biblioref">39</a>]</span> to call variants and then applies a series of complex variant
filtering. As a baseline for the Nanopore data, we use <code>bcftools</code> <span class="citation" data-cites="bXIkkZJW">[<a href="#ref-bXIkkZJW" role="doc-biblioref">40</a>]</span>, with
some filtering of variants to remove low-quality calls and with a mask to avoid
repetitive and structurally variable regions of the Mtb genome.</p>
<h5 id="nanopore-snp-concordance-with-illumina">Nanopore SNP concordance with Illumina</h5>
<p>Whilst the Illumina SNP calls from COMPASS are filtered as part of the pipeline, we had
to settle on filters for the Nanopore SNP calls from <code>bcftools</code>. We used the methodology
from the section
<a href="#comparing-illumina-and-nanopore-snps-to-truth-assemblies">Comparing Illumina and Nanopore SNPs to truth assemblies</a>
to refine the filters in an iterative process. In the end we filter all SNPs with
quality (QUAL column in VCF) below 60, a read position bias less than 0.05, a
segregation-based metric above -0.5, or a variant distance bias below 0.002.</p>
<p>To assess how well the Nanopore SNPs agree with Illumina we first look at SNP
concordance. Two metrics of interest here are the call rate - what proportion of COMPASS
alternate alleles does <code>bcftools</code> make a reference/alternate call - and the concordance
- what proportion of COMPASS alternate alleles does <code>bcftools</code> genotype agree with.
Figure <a href="#fig:alt_concordance">6</a> shows that concordance is very high between the two
technologies, with nearly all samples having a concordance greater than 99.5%. Call rate
is a little lower than this, with the majority of samples being above 97%.</p>
<div id="fig:alt_concordance" class="fignos">
<figure>
<img src="images/alt_concordance.png" alt="" /><figcaption><span>Figure 6:</span> Call rate (Y-axis) and concordance (X-axis) of <code>bcftools</code> SNP calls to Illumina
COMPASS calls. Call rate is what proportion of COMPASS alternate alleles does
<code>bcftools</code> make a reference/alternate call. Concordance is what proportion of COMPASS
alternate alleles does <code>bcftools</code> genotype agree with. Each point represents a sample,
with samples coloured by the site the data came from.</figcaption>
</figure>
</div>
<!--##### Comparing Illumina and Nanopore SNPs to truth assemblies-->
<!--\subsubsection{Basic analysis of per-sample variant calls} \label{sec:evaluateprgs} The-->
<!--aim of this subsection will be to try varying degrees of \prg{} complexity for Mtb-->
<!--sample analysis. At this stage, we have four varieties in mind:-->
<!--\begin{itemize} \item Linear \prg{} - This will just be the standard Mtb reference-->
<!--H37Rv\cite{Camus2002,cole1998,LEW20111}. \item Sparse \prg{} - H37Rv and all variants at-->
<!--frequency above 10\%, from a global Mtb dataset in a paper being prepared in our lab-->
<!--looking at drug resistance prediction. \item Dense \prg{} - The same as Sparse \prg{},-->
<!--but with variants at frequency above 1\%. \item Representative \prg{} - This will-->
<!--contain two high-quality genomes from each lineage, plus all variants from the above-->
<!--dataset at $\geq$5\% frequency. \end{itemize}-->
<!--In all of the above \prg{}s, we will apply the same mask from the baseline analysis and-->
<!--divide the genome into genes and intergenic regions, with a local \prg{} for each.-->
<!--For each of the \prg{}s, we plan to perform the following analysis. Quantify the number-->
<!--of SNPs and indels `pandora` calls per-sample on average and see how this compares to-->
<!--the baseline and truth. We will report on the concordance rate, which is the proportion-->
<!--of shared sites between `pandora` and the truth that agree. Additionally, we will-->
<!--investigate how the complexity of the \prg{} effects the call rates and what the cost in-->
<!--computational performance is.-->
<!--\subsubsection{Reproducing "truth" Illumina clusters using genome graphs} In this-->
<!--section, we will calculate the pairwise SNP distance between all pairs of samples, using-->
<!--the truth set from \hyperref[sec:baseline]{Section \ref{sec:baseline}}. We will then do-->
<!--the same using the `pandora` variant calls from whichever \prg{} in-->
<!--\hyperref[sec:evaluateprgs]{Section \ref{sec:evaluateprgs}} we decide provides the best-->
<!--results. The main figure for this subsection will be a dot plot where each dot is a pair-->
<!--and the X- and Y-axes are the pairwise distance for that pair according to the truth set-->
<!--and `pandora` respectively. In a perfect world, we would expect to see a dead-straight-->
<!--diagonal line from the bottom-left to top-right. In reality, we hope to see a general-->
<!--linear trend. A linear trend would allow us to state that although specific SNP distance-->
<!--thresholds used to define clusters differ between technologies, they are relative and-->
<!--therefore one can use Nanopore data with `pandora` to generate epidemiological clusters-->
<!--for Mtb.-->
<h2 class="page_break_before" id="part-b-training-and-career-development">Part B: Training and career development</h2>
<h3 id="publication-strategy">Publication Strategy</h3>
<blockquote>
<p>Briefly outline the publication strategy; set priorities if needed.</p>
</blockquote>
<h3 id="list-of-publications-papers-in-press-preprints-manuscripts-submittedin-preparation-to-date">List of publications, papers in press, preprints, manuscripts submitted/in preparation to date</h3>
<h3 id="work-plan-and-timeline-for-thesis-submission">Work plan and timeline for thesis submission</h3>
<blockquote>
<p>Provide a work plan to be completed prior to thesis writing and a timeline for submission.</p>
</blockquote>
<h3 id="list-of-scientific-courses-and-conferences-attended-to-date-and-planned-for-next-year">List of scientific courses and conferences attended to date and planned for next year</h3>
<blockquote>
<p>Please list all events in reverse chronological order.</p>
</blockquote>
<h3 id="list-of-additional-training-teaching-and-other-relevant-activities-to-date">List of additional training, teaching and other relevant activities to date</h3>
<blockquote>
<p>Please list all training and other activities in reverse chronological order.</p>
</blockquote>
<h3 id="career-development-plan">Career development plan</h3>
<h4 id="please-describe-your-current-long-term-career-aims-i.e.-3-5-years-after-phd.">Please describe your current long-term career aims (i.e. 3-5 years after PhD).</h4>
<h4 id="please-comment-on-the-types-of-position-you-would-like-to-apply-to-for-after-the-phd-and-your-expected-application-timeline">Please comment on the types of position you would like to apply to for after the PhD and your expected application timeline?</h4>
<blockquote>
<p>If applying for postdoc positions, which fields/fellowships are you considering? If you are applying for non-academic careers, what type of role? Do you have target companies or organisations in mind?</p>
</blockquote>
<h4 id="what-do-you-see-as-your-strengths-2-3-skills">What do you see as your strengths (2-3 skills)?</h4>
<h4 id="what-do-you-see-as-your-areas-for-improvement1-2-3-areas">What do you see as your areas for improvement1 (2-3 areas)?</h4>
<h4 id="what-are-your-career-development-priorities-until-the-end-of-your-contract">What are your career development priorities until the end of your contract?</h4>
<blockquote>
<p>Please take into account both the scientific and non-scientific skills you will need to work on a successful and timely completion of the PhD. List the skills you still need to acquire to achieve your longer-term career aims.</p>
</blockquote>
<h4 id="what-actions-will-you-take-to-develop-these-skills">What actions will you take to develop these skills?</h4>
<blockquote>
<p>This may include training courses, opportunities to practice and develop these skills, or seeking feedback/guidance. N.B. fellows should take 1-2 non-scientific trainings or career development workshops per year, and at least one international conference during their PhD.</p>
</blockquote>
<h2 class="page_break_before" id="part-c-impact-of-covid-19-pandemic">Part C: Impact of COVID-19 pandemic</h2>
<blockquote>
<p>As outlined in the Procedures for COVID-19 related fellows’ contract extensions, students and their GTL should document the impact of the corona crisis on the PhD project. The TAC should assess the approximate project delay and discuss how this could be mitigated over the remaining PhD period. Please use the space below to outline the impact of corona crisis on your project (e.g. lost productivity / project delays).</p>
</blockquote>
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references hanging-indent" role="doc-bibliography">
<div id="ref-owU3wUrb">
<p>1. <strong>Global tuberculosis report 2019</strong> <br />
Organisation mondiale de la santé<br />
(2019) <br />
ISBN: <a href="https://worldcat.org/isbn/9789241565714">9789241565714</a></p>
</div>
<div id="ref-Z3NG0hEI">
<p>2. <strong>Mycobacterial DNA Extraction for Whole-Genome Sequencing from Early Positive Liquid (MGIT) Cultures</strong> <br />
Antonina A. Votintseva, Louise J. Pankhurst, Luke W. Anson, Marcus R. Morgan, Deborah Gascoyne-Binzi, Timothy M. Walker, T. Phuong Quan, David H. Wyllie, Carlos Del Ojo Elias, Mark Wilcox, … Derrick W. Crook<br />
<em>Journal of Clinical Microbiology</em> (2015-04) <a href="https://doi.org/f65dtt">https://doi.org/f65dtt</a> <br />
DOI: <a href="https://doi.org/10.1128/jcm.03073-14">10.1128/jcm.03073-14</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25631807">25631807</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4365189">PMC4365189</a></p>
</div>
<div id="ref-d8RxFwPN">
<p>3. <strong>Rapid antibiotic-resistance predictions from genome sequence data for Staphylococcus aureus and Mycobacterium tuberculosis</strong> <br />
Phelim Bradley, N. Claire Gordon, Timothy M. Walker, Laura Dunn, Simon Heys, Bill Huang, Sarah Earle, Louise J. Pankhurst, Luke Anson, Mariateresa de Cesare, … Zamin Iqbal<br />
<em>Nature Communications</em> (2015-12-21) <a href="https://doi.org/f755tg">https://doi.org/f755tg</a> <br />
DOI: <a href="https://doi.org/10.1038/ncomms10063">10.1038/ncomms10063</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26686880">26686880</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4703848">PMC4703848</a></p>
</div>
<div id="ref-11Jcu87Ae">
<p>4. <strong>Whole-genome sequencing for prediction of Mycobacterium tuberculosis drug susceptibility and resistance: a retrospective cohort study</strong> <br />
Timothy M Walker, Thomas A Kohl, Shaheed V Omar, Jessica Hedge, Carlos Del Ojo Elias, Phelim Bradley, Zamin Iqbal, Silke Feuerriegel, Katherine E Niehaus, Daniel J Wilson, … Tim EA Peto<br />
<em>The Lancet Infectious Diseases</em> (2015-10) <a href="https://doi.org/f3jjtq">https://doi.org/f3jjtq</a> <br />
DOI: <a href="https://doi.org/10.1016/s1473-3099(15)00062-6">10.1016/s1473-3099(15)00062-6</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26116186">26116186</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4579482">PMC4579482</a></p>
</div>
<div id="ref-4zIt0Vce">
<p>5. <strong>Prediction of Susceptibility to First-Line Tuberculosis Drugs by DNA Sequencing</strong> <br />
The CRyPTIC Consortium and the 100,000 Genomes Project<br />
<em>New England Journal of Medicine</em> (2018-10-11) <a href="https://doi.org/d9kj">https://doi.org/d9kj</a> <br />
DOI: <a href="https://doi.org/10.1056/nejmoa1800474">10.1056/nejmoa1800474</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30280646">30280646</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6121966">PMC6121966</a></p>
</div>
<div id="ref-9NLFJZN2">
<p>6. <strong>Same-Day Diagnostic and Surveillance Data for Tuberculosis via Whole-Genome Sequencing of Direct Respiratory Samples</strong> <br />
Antonina A. Votintseva, Phelim Bradley, Louise Pankhurst, Carlos del Ojo Elias, Matthew Loose, Kayzad Nilgiriwala, Anirvan Chatterjee, E. Grace Smith, Nicolas Sanderson, Timothy M. Walker, … Zamin Iqbal<br />
<em>Journal of Clinical Microbiology</em> (2017-05) <a href="https://doi.org/f94vt4">https://doi.org/f94vt4</a> <br />
DOI: <a href="https://doi.org/10.1128/jcm.02483-16">10.1128/jcm.02483-16</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28275074">28275074</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5405248">PMC5405248</a></p>
</div>
<div id="ref-5WDPm9pG">
<p>7. <strong>Why prokaryotes have pangenomes</strong> <br />
James O. McInerney, Alan McNally, Mary J. O’Connell<br />
<em>Nature Microbiology</em> (2017-03-28) <a href="https://doi.org/gfw8gq">https://doi.org/gfw8gq</a> <br />
DOI: <a href="https://doi.org/10.1038/nmicrobiol.2017.40">10.1038/nmicrobiol.2017.40</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28350002">28350002</a></p>
</div>
<div id="ref-IYW3CNeu">
<p>8. <strong>panX: pan-genome analysis and exploration</strong> <br />
Wei Ding, Franz Baumdicker, Richard A Neher<br />
<em>Nucleic Acids Research</em> (2018-01-09) <a href="https://doi.org/gczkbr">https://doi.org/gczkbr</a> <br />
DOI: <a href="https://doi.org/10.1093/nar/gkx977">10.1093/nar/gkx977</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29077859">29077859</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5758898">PMC5758898</a></p>
</div>
<div id="ref-QrKrCCsB">
<p>9. <strong>BulkVis: a graphical viewer for Oxford nanopore bulk FAST5 files</strong> <br />
Alexander Payne, Nadine Holmes, Vardhman Rakyan, Matthew Loose<br />
<em>Bioinformatics</em> (2019-07-01) <a href="https://doi.org/gfnkxj">https://doi.org/gfnkxj</a> <br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/bty841">10.1093/bioinformatics/bty841</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30462145">30462145</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6596899">PMC6596899</a></p>
</div>
<div id="ref-lwFwJNS9">
<p>10. <strong>Performance of neural network basecalling tools for Oxford Nanopore sequencing</strong> <br />
Ryan R. Wick, Louise M. Judd, Kathryn E. Holt<br />
<em>Genome Biology</em> (2019-06-24) <a href="https://doi.org/gf4jwm">https://doi.org/gf4jwm</a> <br />
DOI: <a href="https://doi.org/10.1186/s13059-019-1727-y">10.1186/s13059-019-1727-y</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31234903">31234903</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6591954">PMC6591954</a></p>
</div>
<div id="ref-Oe1xiQra">
<p>11. <strong>Real-time, portable genome sequencing for Ebola surveillance</strong> <br />
Joshua Quick, Nicholas J. Loman, Sophie Duraffour, Jared T. Simpson, Ettore Severi, Lauren Cowley, Joseph Akoi Bore, Raymond Koundouno, Gytis Dudas, Amy Mikhail, … Miles W. Carroll<br />
<em>Nature</em> (2016-02-03) <a href="https://doi.org/f88652">https://doi.org/f88652</a> <br />
DOI: <a href="https://doi.org/10.1038/nature16996">10.1038/nature16996</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26840485">26840485</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4817224">PMC4817224</a></p>
</div>
<div id="ref-LurPKJ01">
<p>12. <strong>Exploring the limit of using a deep neural network on pileup data for germline variant calling</strong> <br />
Ruibang Luo, Chak-Lim Wong, Yat-Sing Wong, Chi-Ian Tang, Chi-Man Liu, Chi-Ming Leung, Tak-Wah Lam<br />
<em>Nature Machine Intelligence</em> (2020-04-06) <a href="https://doi.org/d9kq">https://doi.org/d9kq</a> <br />
DOI: <a href="https://doi.org/10.1038/s42256-020-0167-4">10.1038/s42256-020-0167-4</a></p>
</div>
<div id="ref-VSoEWPFn">
<p>13. <strong>nanoporetech/medaka</strong> <br />
GitHub<br />
<a href="https://github.com/nanoporetech/medaka">https://github.com/nanoporetech/medaka</a></p>
</div>
<div id="ref-pPxMDtt0">
<p>14. <strong>High precision <em>Neisseria gonorrhoeae</em> variant and antimicrobial resistance calling from metagenomic Nanopore sequencing</strong> <br />
Nicholas D. Sanderson, Jeremy Swann, Leanne Barker, James Kavanagh, Sarah Hoosdally, Derrick Crook, Teresa L. Street, David W. Eyre, The GonFast Investigators Group<br />
<em>Genome Research</em> (2020-09) <a href="https://doi.org/ghcbjr">https://doi.org/ghcbjr</a> <br />
DOI: <a href="https://doi.org/10.1101/gr.262865.120">10.1101/gr.262865.120</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32873606">32873606</a></p>
</div>
<div id="ref-19hwdhKN">
<p>15. <strong>Reducing storage requirements for biological sequence comparison</strong> <br />
M. Roberts, W. Hayes, B. R. Hunt, S. M. Mount, J. A. Yorke<br />
<em>Bioinformatics</em> (2004-07-15) <a href="https://doi.org/dkhs8w">https://doi.org/dkhs8w</a> <br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/bth408">10.1093/bioinformatics/bth408</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/15256412">15256412</a></p>
</div>
<div id="ref-1DgB6gqeS">
<p>16. <strong>Minimap and miniasm: fast mapping and de novo assembly for noisy long sequences</strong> <br />
Heng Li<br />
<em>Bioinformatics</em> (2016-07-15) <a href="https://doi.org/f8zxc3">https://doi.org/f8zxc3</a> <br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btw152">10.1093/bioinformatics/btw152</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27153593">27153593</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4937194">PMC4937194</a></p>
</div>
<div id="ref-2FYPCKEv">
<p>17. <strong>A Natural Encoding of Genetic Variation in a Burrows-Wheeler Transform to Enable Mapping and Genome Inference</strong> <br />
Sorina Maciuca, Carlos del Ojo Elias, Gil McVean, Zamin Iqbal<br />
<em>Lecture Notes in Computer Science</em> (2016) <a href="https://doi.org/d9ks">https://doi.org/d9ks</a> <br />
DOI: <a href="https://doi.org/10.1007/978-3-319-43681-4_18">10.1007/978-3-319-43681-4_18</a></p>
</div>
<div id="ref-7AwGVQXO">
<p>18. <strong>Improved genome inference in the MHC using a population reference graph</strong> <br />
Alexander Dilthey, Charles Cox, Zamin Iqbal, Matthew R Nelson, Gil McVean<br />
<em>Nature Genetics</em> (2015-04-27) <a href="https://doi.org/f7dt83">https://doi.org/f7dt83</a> <br />
DOI: <a href="https://doi.org/10.1038/ng.3257">10.1038/ng.3257</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25915597">25915597</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4449272">PMC4449272</a></p>
</div>
<div id="ref-h4yvs7gl">
<p>19. <strong>Genetic variation: molecular mechanisms and impact on microbial evolution</strong> <br />
Werner Arber<br />
<em>FEMS Microbiology Reviews</em> (2000-01) <a href="https://doi.org/b4x7zf">https://doi.org/b4x7zf</a> <br />
DOI: <a href="https://doi.org/10.1111/j.1574-6976.2000.tb00529.x">10.1111/j.1574-6976.2000.tb00529.x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/10640595">10640595</a></p>
</div>
<div id="ref-QdpEzyJH">
<p>20. <strong>CFSAN SNP Pipeline: an automated method for constructing SNP matrices from next-generation sequence data</strong> <br />
Steve Davis, James B. Pettengill, Yan Luo, Justin Payne, Al Shpuntoff, Hugh Rand, Errol Strain<br />
<em>PeerJ Computer Science</em> (2015-08-26) <a href="https://doi.org/d9np">https://doi.org/d9np</a> <br />
DOI: <a href="https://doi.org/10.7717/peerj-cs.20">10.7717/peerj-cs.20</a></p>
</div>
<div id="ref-Fp8BI1Lc">
<p>21. <strong>NanoSim: nanopore sequence read simulator based on statistical characterization</strong> <br />
Chen Yang, Justin Chu, René L Warren, Inanç Birol<br />
<em>GigaScience</em> (2017-04) <a href="https://doi.org/gbj69k">https://doi.org/gbj69k</a> <br />
DOI: <a href="https://doi.org/10.1093/gigascience/gix010">10.1093/gigascience/gix010</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28327957">28327957</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5530317">PMC5530317</a></p>
</div>
<div id="ref-7m2CKnnh">
<p>22. <strong>Karel-Brinda/Nanosim-H: Nanosim-H 1.1.0.4</strong> <br />
Karel Břinda, Chen Yang, Justin Chu, Jasper Linthorst, Wiktor Franus<br />
<em>Zenodo</em> (2018-08-07) <a href="https://doi.org/d9nq">https://doi.org/d9nq</a> <br />
DOI: <a href="https://doi.org/10.5281/zenodo.1341250">10.5281/zenodo.1341250</a></p>
</div>
<div id="ref-gmH6YDca">
<p>23. <strong>Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM</strong> <br />
Heng Li<br />
<em>arXiv</em> (2013-05-28) <a href="https://arxiv.org/abs/1303.3997">https://arxiv.org/abs/1303.3997</a></p>
</div>
<div id="ref-6S74PtQr">
<p>24. <strong>tseemann/snippy</strong> <br />
GitHub<br />
<a href="https://github.com/tseemann/snippy">https://github.com/tseemann/snippy</a></p>
</div>
<div id="ref-NcYZqBux">
<p>25. <strong>Snakemake–a scalable bioinformatics workflow engine</strong> <br />
J. Koster, S. Rahmann<br />
<em>Bioinformatics</em> (2012-08-20) <a href="https://doi.org/gd2xzq">https://doi.org/gd2xzq</a> <br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/bts480">10.1093/bioinformatics/bts480</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22908215">22908215</a></p>
</div>
<div id="ref-172cJaw4Q">
<p>26. <strong>Minimap2: pairwise alignment for nucleotide sequences</strong> <br />
Heng Li<br />
<em>Bioinformatics</em> (2018-09-15) <a href="https://doi.org/gdhbqt">https://doi.org/gdhbqt</a> <br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/bty191">10.1093/bioinformatics/bty191</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29750242">29750242</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6137996">PMC6137996</a></p>
</div>
<div id="ref-4U7AuMzS">
<p>27. <strong>Rasusa: Randomly subsample sequencing reads to a specified coverage</strong> <br />
Michael Hall<br />
<em>Zenodo</em> (2020-03-27) <a href="https://doi.org/d9rz">https://doi.org/d9rz</a> <br />
DOI: <a href="https://doi.org/10.5281/zenodo.3731394">10.5281/zenodo.3731394</a></p>
</div>
<div id="ref-c17o0GrE">
<p>28. <strong>Evolutionary pathway analysis and unified classification of East Asian lineage of Mycobacterium tuberculosis</strong> <br />
Egor Shitikov, Sergey Kolchenko, Igor Mokrousov, Julia Bespyatykh, Dmitry Ischenko, Elena Ilina, Vadim Govorun<br />
<em>Scientific Reports</em> (2017-08-23) <a href="https://doi.org/gbvbxh">https://doi.org/gbvbxh</a> <br />
DOI: <a href="https://doi.org/10.1038/s41598-017-10018-5">10.1038/s41598-017-10018-5</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28835627">28835627</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5569047">PMC5569047</a></p>
</div>
<div id="ref-6J5Bt0f7">
<p>29. <strong>Multiple Introductions of Mycobacterium tuberculosis Lineage 2–Beijing Into Africa Over Centuries</strong> <br />
Liliana K. Rutaihwa, Fabrizio Menardo, David Stucki, Sebastian M. Gygli, Serej D. Ley, Bijaya Malla, Julia Feldmann, Sonia Borrell, Christian Beisel, Kerren Middelkoop, … Sebastien Gagneux<br />
<em>Frontiers in Ecology and Evolution</em> (2019-04-16) <a href="https://doi.org/d9r2">https://doi.org/d9r2</a> <br />
DOI: <a href="https://doi.org/10.3389/fevo.2019.00112">10.3389/fevo.2019.00112</a></p>
</div>
<div id="ref-1CS78SxuA">
<p>30. <strong>Mycobacterium tuberculosis lineage 4 comprises globally distributed and geographically restricted sublineages</strong> <br />
David Stucki, Daniela Brites, Leïla Jeljeli, Mireia Coscolla, Qingyun Liu, Andrej Trauner, Lukas Fenner, Liliana Rutaihwa, Sonia Borrell, Tao Luo, … Sebastien Gagneux<br />
<em>Nature Genetics</em> (2016-10-31) <a href="https://doi.org/f9dg9j">https://doi.org/f9dg9j</a> <br />
DOI: <a href="https://doi.org/10.1038/ng.3704">10.1038/ng.3704</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27798628">27798628</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5238942">PMC5238942</a></p>
</div>
<div id="ref-tGyXVFvJ">
<p>31. <strong>Beyond the SNP Threshold: Identifying Outbreak Clusters Using Inferred Transmissions</strong> <br />
James Stimson, Jennifer Gardy, Barun Mathema, Valeriu Crudu, Ted Cohen, Caroline Colijn<br />
<em>Molecular Biology and Evolution</em> (2019-03) <a href="https://doi.org/d9r7">https://doi.org/d9r7</a> <br />
DOI: <a href="https://doi.org/10.1093/molbev/msy242">10.1093/molbev/msy242</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30690464">30690464</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6389316">PMC6389316</a></p>
</div>
<div id="ref-64Rp8zLg">
<p>32. <strong>Interpreting whole genome sequencing for investigating tuberculosis transmission: a systematic review</strong> <br />
Hollie-Ann Hatherell, Caroline Colijn, Helen R. Stagg, Charlotte Jackson, Joanne R. Winter, Ibrahim Abubakar<br />
<em>BMC Medicine</em> (2016-03-23) <a href="https://doi.org/f8gsk2">https://doi.org/f8gsk2</a> <br />
DOI: <a href="https://doi.org/10.1186/s12916-016-0566-x">10.1186/s12916-016-0566-x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27005433">27005433</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4804562">PMC4804562</a></p>
</div>
<div id="ref-18CTtAeF9">
<p>33. <strong>Whole-genome sequencing to delineate Mycobacterium tuberculosis outbreaks: a retrospective observational study</strong> <br />
Timothy M Walker, Camilla LC Ip, Ruth H Harrell, Jason T Evans, Georgia Kapatai, Martin J Dedicoat, David W Eyre, Daniel J Wilson, Peter M Hawkey, Derrick W Crook, … Tim EA Peto<br />
<em>The Lancet Infectious Diseases</em> (2013-02) <a href="https://doi.org/f2g6zn">https://doi.org/f2g6zn</a> <br />
DOI: <a href="https://doi.org/10.1016/s1473-3099(12)70277-3">10.1016/s1473-3099(12)70277-3</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23158499">23158499</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3556524">PMC3556524</a></p>
</div>
<div id="ref-dRRsEE8Z">
<p>34. <strong>Bottlenecks and broomsticks: the molecular evolution of Mycobacterium bovis</strong> <br />
Noel H. Smith, Stephen V. Gordon, Ricardo de la Rua-Domenech, Richard S. Clifton-Hadley, R. Glyn Hewinson<br />
<em>Nature Reviews Microbiology</em> (2006-09) <a href="https://doi.org/fhqqkv">https://doi.org/fhqqkv</a> <br />
DOI: <a href="https://doi.org/10.1038/nrmicro1472">10.1038/nrmicro1472</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/16912712">16912712</a></p>
</div>
<div id="ref-hpSc80yh">
<p>35. <strong>Evidence for Recombination in Mycobacterium tuberculosis</strong> <br />
Xiaoming Liu, Michaela M. Gutacker, James M. Musser, Yun-Xin Fu<br />
<em>Journal of Bacteriology</em> (2006-12-01) <a href="https://doi.org/ftp6r2">https://doi.org/ftp6r2</a> <br />
DOI: <a href="https://doi.org/10.1128/jb.01062-06">10.1128/jb.01062-06</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/16997954">16997954</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1698211">PMC1698211</a></p>
</div>
<div id="ref-1H760aM5J">
<p>36. <strong>Genomic analysis of smooth tubercle bacilli provides insights into ancestry and pathoadaptation of Mycobacterium tuberculosis</strong> <br />
Philip Supply, Michael Marceau, Sophie Mangenot, David Roche, Carine Rouanet, Varun Khanna, Laleh Majlessi, Alexis Criscuolo, Julien Tap, Alexandre Pawlik, … Roland Brosch<br />
<em>Nature Genetics</em> (2013-01-06) <a href="https://doi.org/f4mrqv">https://doi.org/f4mrqv</a> <br />
DOI: <a href="https://doi.org/10.1038/ng.2517">10.1038/ng.2517</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23291586">23291586</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3856870">PMC3856870</a></p>
</div>
<div id="ref-Coyn9XyF">
<p>37. <strong>Computational genomics-proteomics and Phylogeny analysis of twenty one mycobacterial genomes (Tuberculosis &amp; non Tuberculosis strains)</strong> <br />
Fathiah Zakham, Othmane Aouane, David Ussery, Abdelaziz Benjouad, Moulay Ennaji<br />
<em>Microbial Informatics and Experimentation</em> (2012) <a href="https://doi.org/d9r8">https://doi.org/d9r8</a> <br />
DOI: <a href="https://doi.org/10.1186/2042-5783-2-7">10.1186/2042-5783-2-7</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22929624">22929624</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3504576">PMC3504576</a></p>
</div>
<div id="ref-Lvhz1sqs">
<p>38. <strong>Towards standardisation: comparison of five whole genome sequencing (WGS) analysis pipelines for detection of epidemiologically linked tuberculosis cases</strong> <br />
Rana Jajou, Thomas A Kohl, Timothy Walker, Anders Norman, Daniela Maria Cirillo, Elisa Tagliani, Stefan Niemann, Albert de Neeling, Troels Lillebaek, Richard M Anthony, Dick van Soolingen<br />
<em>Eurosurveillance</em> (2019-12-12) <a href="https://doi.org/d9r9">https://doi.org/d9r9</a> <br />
DOI: <a href="https://doi.org/10.2807/1560-7917.es.2019.24.50.1900130">10.2807/1560-7917.es.2019.24.50.1900130</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31847944">31847944</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6918587">PMC6918587</a></p>
</div>
<div id="ref-IPc8cy7s">
<p>39. <strong>The Sequence Alignment/Map format and SAMtools</strong> <br />
H. Li, B. Handsaker, A. Wysoker, T. Fennell, J. Ruan, N. Homer, G. Marth, G. Abecasis, R. Durbin, 1000 Genome Project Data Processing Subgroup<br />
<em>Bioinformatics</em> (2009-06-08) <a href="https://doi.org/ff6426">https://doi.org/ff6426</a> <br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btp352">10.1093/bioinformatics/btp352</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19505943">19505943</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2723002">PMC2723002</a></p>
</div>
<div id="ref-bXIkkZJW">
<p>40. <strong>Improving SNP discovery by base alignment quality</strong> <br />
H. Li<br />
<em>Bioinformatics</em> (2011-02-13) <a href="https://doi.org/fw7k5k">https://doi.org/fw7k5k</a> <br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btr076">10.1093/bioinformatics/btr076</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/21320865">21320865</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3072548">PMC3072548</a></p>
</div>
</div>
<!-- default theme -->

<style>
    /* import google fonts */
    @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
    @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

    /* -------------------------------------------------- */
    /* global */
    /* -------------------------------------------------- */

    /* all elements */
    * {
        /* force sans-serif font unless specified otherwise */
        font-family: "Open Sans", "Helvetica", sans-serif;

        /* prevent text inflation on some mobile browsers */
        -webkit-text-size-adjust: none !important;
        -moz-text-size-adjust: none !important;
        -o-text-size-adjust: none !important;
        text-size-adjust: none !important;
    }

    @media only screen {
        /* "page" element */
        body {
            position: relative;
            box-sizing: border-box;
            font-size: 12pt;
            line-height: 1.5;
            max-width: 8.5in;
            margin: 20px auto;
            padding: 40px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* "page" element */
        body {
            padding: 20px;
            margin: 0;
            border-radius: 0;
            border: none;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
            background: none;
        }
    }

    /* -------------------------------------------------- */
    /* headings */
    /* -------------------------------------------------- */

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
        margin: 20px 0;
        padding: 0;
        font-weight: bold;
    }

    /* biggest heading */
    h1 {
        margin: 40px 0;
        text-align: center;
    }

    /* second biggest heading */
    h2 {
        margin-top: 30px;
        padding-bottom: 5px;
        border-bottom: solid 1px #bdbdbd;
    }

    /* heading font sizes */
    h1 {
        font-size: 2em;
    }
    h2 {
        font-size: 1.5em;
    }
    h3{
        font-size: 1.35em;
    }
    h4 {
        font-size: 1.25em;
    }
    h5 {
        font-size: 1.15em;
    }
    h6 {
        font-size: 1em;
    }

    /* -------------------------------------------------- */
    /* manuscript header */
    /* -------------------------------------------------- */

    /* manuscript title */
    header > h1 {
        margin: 0;
    }

    /* manuscript title caption text (ie "automatically generated on") */
    header + p {
        text-align: center;
        margin-top: 10px;
    }

    /* -------------------------------------------------- */
    /* text elements */
    /* -------------------------------------------------- */

    /* links */
    a {
        color: #2196f3;
        overflow-wrap: break-word;
    }

    /* normal links (not empty, not button link, not syntax highlighting link) */
    a:not(:empty):not(.button):not(.sourceLine) {
        padding-left: 1px;
        padding-right: 1px;
    }

    /* superscripts and subscripts */
    sub,
    sup {
        /* prevent from affecting line height */
        line-height: 0;
    }

    /* unordered and ordered lists*/
    ul,
    ol {
        padding-left: 20px;
    }

    /* class for styling text semibold */
    .semibold {
        font-weight: 600;
    }

    /* class for styling elements horizontally left aligned */
    .left {
        display: block;
        text-align: left;
        margin-left: auto;
        margin-right: 0;
        justify-content: left;
    }

    /* class for styling elements horizontally centered */
    .center {
        display: block;
        text-align: center;
        margin-left: auto;
        margin-right: auto;
        justify-content: center;
    }

    /* class for styling elements horizontally right aligned */
    .right {
        display: block;
        text-align: right;
        margin-left: 0;
        margin-right: auto;
        justify-content: right;
    }

    /* -------------------------------------------------- */
    /* section elements */
    /* -------------------------------------------------- */

    /* horizontal divider line */
    hr {
        border: none;
        height: 1px;
        background: #bdbdbd;
    }

    /* paragraphs, horizontal dividers, figures, tables, code */
    p,
    hr,
    figure,
    table,
    pre {
        /* treat all as "paragraphs", with consistent vertical margins */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* figures */
    /* -------------------------------------------------- */

    /* figure */
    figure {
        max-width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure caption */
    figcaption {
        padding: 0;
        padding-top: 10px;
    }

    /* figure image element */
    figure img {
        max-width: 100%;
        display: block;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure auto-number */
    img + figcaption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* tables */
    /* -------------------------------------------------- */

    /* table */
    table {
        border-collapse: collapse;
        border-spacing: 0;
        width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* table cells */
    th,
    td {
        border: solid 1px #bdbdbd;
        padding: 10px;
        /* squash table if too wide for page by forcing line breaks */
        overflow-wrap: break-word;
        word-break: break-word;
    }

    /* header row and even rows */
    th,
    tr:nth-child(2n) {
        background-color: #fafafa;
    }

    /* odd rows */
    tr:nth-child(2n + 1) {
        background-color: #ffffff;
    }

    /* table caption */
    caption {
        text-align: left;
        padding: 0;
        padding-bottom: 10px;
    }

    /* table auto-number */
    table > caption > span:first-of-type,
    div.table_wrapper > table > caption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* code */
    /* -------------------------------------------------- */

    /* multi-line code block */
    pre {
        padding: 10px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
        break-inside: avoid;
        text-align: left;
    }

    /* inline code, ie code within normal text */
    :not(pre) > code {
        padding: 0 4px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
    }

    /* code text */
    /* apply all children, to reach syntax highlighting sub-elements */
    code,
    code * {
        /* force monospace font */
        font-family: "Source Code Pro", "Courier New", monospace;
    }

    /* -------------------------------------------------- */
    /* quotes */
    /* -------------------------------------------------- */

    /* quoted text */
    blockquote {
        margin: 0;
        padding: 0;
        border-left: 4px solid #bdbdbd;
        padding-left: 16px;
        break-inside: avoid;
    }

    /* -------------------------------------------------- */
    /* banners */
    /* -------------------------------------------------- */

    /* info banners */
    .banner {
        box-sizing: border-box;
        display: block;
        position: relative;
        width: 100%;
        margin-top: 20px;
        margin-bottom: 20px;
        padding: 20px;
        text-align: center;
    }

    /* paragraph in banner */
    .banner > p {
        margin: 0;
    }

    /* -------------------------------------------------- */
    /* highlight colors */
    /* -------------------------------------------------- */

    .white {
        background: #ffffff;
    }
    .lightgrey {
        background: #eeeeee;
    }
    .grey {
        background: #757575;
    }
    .darkgrey {
        background: #424242;
    }
    .black {
        background: #000000;
    }
    .lightred {
        background: #ffcdd2;
    }
    .lightyellow {
        background: #ffecb3;
    }
    .lightgreen {
        background: #dcedc8;
    }
    .lightblue {
        background: #e3f2fd;
    }
    .lightpurple {
        background: #f3e5f5;
    }
    .red {
        background: #f44336;
    }
    .orange {
        background: #ff9800;
    }
    .yellow {
        background: #ffeb3b;
    }
    .green {
        background: #4caf50;
    }
    .blue {
        background: #2196f3;
    }
    .purple {
        background: #9c27b0;
    }
    .white,
    .lightgrey,
    .lightred,
    .lightyellow,
    .lightgreen,
    .lightblue,
    .lightpurple,
    .orange,
    .yellow,
    .white a,
    .lightgrey a,
    .lightred a,
    .lightyellow a,
    .lightgreen a,
    .lightblue a,
    .lightpurple a,
    .orange a,
    .yellow a {
        color: #000000;
    }
    .grey,
    .darkgrey,
    .black,
    .red,
    .green,
    .blue,
    .purple,
    .grey a,
    .darkgrey a,
    .black a,
    .red a,
    .green a,
    .blue a,
    .purple a {
        color: #ffffff;
    }

    /* -------------------------------------------------- */
    /* buttons */
    /* -------------------------------------------------- */

    /* class for styling links like buttons */
    .button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        margin: 5px;
        padding: 10px 20px;
        font-size: 0.75em;
        font-weight: 600;
        text-transform: uppercase;
        text-decoration: none;
        letter-spacing: 1px;
        background: none;
        color: #2196f3;
        border: solid 1px #bdbdbd;
        border-radius: 5px;
    }

    /* buttons when hovered */
    .button:hover:not([disabled]),
    .icon_button:hover:not([disabled]) {
        cursor: pointer;
        background: #f5f5f5;
    }

    /* buttons when disabled */
    .button[disabled],
    .icon_button[disabled] {
        opacity: 0.35;
        pointer-events: none;
    }

    /* class for styling buttons containg only single icon */
    .icon_button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        text-decoration: none;
        margin: 0;
        padding: 0;
        background: none;
        border-radius: 5px;
        border: none;
        width: 20px;
        height: 20px;
        min-width: 20px;
        min-height: 20px;
    }

    /* icon button inner svg image */
    .icon_button > svg {
        height: 16px;
    }

    /* -------------------------------------------------- */
    /* icons */
    /* -------------------------------------------------- */

    /* class for styling icons inline with text */
    .inline_icon {
        height: 1em;
        position: relative;
        top: 0.125em;
    }

    /* -------------------------------------------------- */
    /* print control */
    /* -------------------------------------------------- */

    @media print {
        @page {
            /* suggested printing margin */
            margin: 0.5in;
        }

        /* document and "page" elements */
        html, body {
            margin: 0;
            padding: 0;
            width: 100%;
            height: 100%;
        }

        /* "page" element */
        body {
            font-size: 11pt !important;
            line-height: 1.35;
        }

        /* all headings */
        h1,
        h2,
        h3,
        h4,
        h5,
        h6 {
            margin: 15px 0;
        }

        /* figures and tables */
        figure, table {
            font-size: 0.85em;
        }

        /* table cells */
        th,
        td {
            padding: 5px;
        }

        /* shrink font awesome icons */
        i.fas,
        i.fab,
        i.far,
        i.fal {
            transform: scale(0.85);
        }

        /* decrease banner margins */
        .banner {
            margin-top: 15px;
            margin-bottom: 15px;
            padding: 15px;
        }

        /* class for centering an element vertically on its own page */
        .page_center {
            margin: auto;
            width: 100%;
            height: 100%;
            display: flex;
            align-items: center;
            vertical-align: middle;
            break-before: page;
            break-after: page;
        }

        /* always insert a page break before the element */
        .page_break_before {
            break-before: page;
        }

        /* always insert a page break after the element */
        .page_break_after {
            break-after: page;
        }

        /* avoid page break before the element */
        .page_break_before_avoid {
            break-before: avoid;
        }

        /* avoid page break after the element */
        .page_break_after_avoid {
            break-after: avoid;
        }

        /* avoid page break inside the element */
        .page_break_inside_avoid {
            break-inside: avoid;
        }
    }

    /* -------------------------------------------------- */
    /* override pandoc css quirks */
    /* -------------------------------------------------- */

    .sourceCode {
        /* prevent unsightly overflow in wide code blocks */
        overflow: auto !important;
    }

    div.sourceCode {
        /* prevent background fill on top-most code block  container */
        background: none !important;
    }

    .sourceCode * {
        /* force consistent line spacing */
        line-height: 1.5 !important;
    }

    div.sourceCode {
        /* style code block margins same as <pre> element */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* tablenos */
    /* -------------------------------------------------- */

    /* tablenos wrapper */
    .tablenos {
        /* show scrollbar on tables if necessary to prevent overflow */
        width: 100%;
        margin: 20px 0;
    }

    .tablenos > table {
        /* move margins from table to table_wrapper to allow margin collapsing */
        margin: 0;
    }

    @media only screen {
        /* tablenos wrapper */
        .tablenos {
            /* show scrollbar on tables if necessary to prevent overflow */
            overflow-x: auto !important;
        }

        .tablenos th,
        .tablenos td {
            overflow-wrap: unset !important;
            word-break: unset !important;
        }

        /* table in wrapper */
        .tablenos table,
        .tablenos table * {
            /* don't break table words */
            overflow-wrap: normal !important;
        }
    }

    /* -------------------------------------------------- */
    /* mathjax */
    /* -------------------------------------------------- */

    /* mathjax containers */
    .math.display > span:not(.MathJax_Preview) {
        /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
        display: flex !important;
        overflow-x: auto !important;
        overflow-y: hidden !important;
        justify-content: center;
        align-items: center;
        margin: 0 !important;
    }

    /* right click menu */
    .MathJax_Menu {
        border-radius: 5px !important;
        border: solid 1px #bdbdbd !important;
        box-shadow: none !important;
    }

    /* equation auto-number */
    span[id^="eq:"] > span.math.display + span {
        font-weight: 600;
    }

    /* equation */
    span[id^="eq:"] > span.math.display > span {
        /* nudge to make room for equation auto-number and anchor */
        margin-right: 60px !important;
    }

    /* -------------------------------------------------- */
    /* anchors plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anchor button */
        .anchor {
            opacity: 0;
            margin-left: 5px;
        }

        /* anchor buttons within <h2>'s */
        h2 .anchor {
            margin-left: 10px;
        }

        /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
        *:hover > .anchor,
        .anchor:hover,
        .anchor:focus {
            opacity: 1;
        }

        /* anchor button when hovered */
        .anchor:hover {
            cursor: pointer;
        }
    }

    /* always show anchor button on devices with no mouse/hover ability */
    @media (hover: none) {
        .anchor {
            opacity: 1;
        }
    }

    /* always hide anchor button on print */
    @media only print {
        .anchor {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* accordion plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* accordion arrow button */
        .accordion_arrow {
            margin-right: 10px;
        }

        /* arrow icon when <h2> data-collapsed attribute true */
        h2[data-collapsed="true"] > .accordion_arrow > svg {
            transform: rotate(-90deg);
        }

        /* all elements (except <h2>'s) when data-collapsed attribute true */
        *:not(h2)[data-collapsed="true"] {
            display: none;
        }

        /* accordion arrow button when hovered and <h2>'s when hovered */
        .accordion_arrow:hover,
        h2[data-collapsed="true"]:hover,
        h2[data-collapsed="false"]:hover {
            cursor: pointer;
        }
    }

    /* always hide accordion arrow button on print */
    @media only print {
        .accordion_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* tooltips plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* tooltip container */
        #tooltip {
            position: absolute;
            width: 50%;
            min-width: 240px;
            max-width: 75%;
            z-index: 1;
        }

        /* tooltip content */
        #tooltip_content {
            margin-bottom: 5px;
            padding: 20px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
            overflow-wrap: break-word;
        }

        /* tooltip copy of paragraphs and figures */
        #tooltip_content > p,
        #tooltip_content > figure {
            margin: 0;
            max-height: 320px;
            overflow-y: auto;
        }

        /* tooltip copy of <img> */
        #tooltip_content > figure > img {
            max-height: 260px;
        }

        /* navigation bar */
        #tooltip_nav_bar {
            margin-top: 10px;
            text-align: center;
        }

        /* navigation bar previous/next buton */
        #tooltip_nav_bar > .icon_button {
            position: relative;
            top: 3px;
        }

        /* navigation bar previous button */
        #tooltip_nav_bar > .icon_button:first-of-type {
            margin-right: 5px;
        }

        /* navigation bar next button */
        #tooltip_nav_bar > .icon_button:last-of-type {
            margin-left: 5px;
        }
    }

    /* always hide tooltip on print */
    @media only print {
        #tooltip {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* jump to first plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* jump button */
        .jump_arrow {
            position: relative;
            top: 0.125em;
            margin-right: 5px;
        }
    }

    /* always hide jump button on print */
    @media only print {
        .jump_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* link highlight plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anything with data-highlighted attribute true */
        [data-highlighted="true"] {
            background: #ffeb3b;
        }

        /* anything with data-selected attribute true */
        [data-selected="true"] {
            background: #ff8a65 !important;
        }

        /* animation definition for glow */
        @keyframes highlight_glow {
            0% {
                background: none;
            }
            10% {
                background: #bbdefb;
            }
            100% {
                background: none;
            }
        }

        /* anything with data-glow attribute true */
        [data-glow="true"] {
            animation: highlight_glow 2s;
        }
    }

    /* -------------------------------------------------- */
    /* table of contents plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* toc panel */
        #toc_panel {
            box-sizing: border-box;
            position: fixed;
            top: 0;
            left: 0;
            background: #ffffff;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            z-index: 2;
        }

        /* toc panel when closed */
        #toc_panel[data-open="false"] {
            min-width: 60px;
            width: 60px;
            height: 60px;
            border-right: solid 1px #bdbdbd;
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc panel when open */
        #toc_panel[data-open="true"] {
            min-width: 260px;
            max-width: 480px;
            /* keep panel edge consistent distance away from "page" edge */
            width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
            bottom: 0;
            border-right: solid 1px #bdbdbd;
        }

        /* toc panel header */
        #toc_header {
            box-sizing: border-box;
            display: flex;
            flex-direction: row;
            align-items: center;
            height: 60px;
            margin: 0;
            padding: 20px;
        }

        /* toc panel header when hovered */
        #toc_header:hover {
            cursor: pointer;
        }

        /* toc panel header when panel open */
        #toc_panel[data-open="true"] > #toc_header {
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc open/close header button */
        #toc_button {
            margin-right: 20px;
        }

        /* hide toc list and header text when closed */
        #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
        #toc_panel[data-open="false"] > #toc_list {
            display: none;
        }

        /* toc list of entries */
        #toc_list {
            box-sizing: border-box;
            width: 100%;
            padding: 20px;
            position: absolute;
            top: calc(60px + 1px);
            bottom: 0;
            overflow: auto;
        }

        /* toc entry, link to section in document */
        .toc_link {
            display: block;
            padding: 5px;
            position: relative;
            font-weight: 600;
            text-decoration: none;
        }

        /* toc entry when hovered or when "viewed" */
        .toc_link:hover,
        .toc_link[data-viewing="true"] {
            background: #f5f5f5;
        }

        /* toc entry, level 1 indentation */
        .toc_link[data-level="1"] {
            margin-left: 0;
        }

        /* toc entry, level 2 indentation */
        .toc_link[data-level="2"] {
            margin-left: 20px;
        }

        /* toc entry, level 3 indentation */
        .toc_link[data-level="3"] {
            margin-left: 40px;
        }

        /* toc entry, level 4 indentation */
        .toc_link[data-level="4"] {
            margin-left: 60px;
        }

        /* toc entry bullets */
        #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
            position: absolute;
            left: -15px;
            top: -1px;
            font-size: 1.5em;
        }

        /* toc entry, level 2 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
            content: "\2022";
        }

        /* toc entry, level 3 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
            content: "\25AB";
        }

        /* toc entry, level 4 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
            content: "-";
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* push <body> ("page") element down to make room for toc icon */
        .toc_body_nudge {
            padding-top: 60px;
        }

        /* toc icon when panel closed and not hovered */
        #toc_panel[data-open="false"]:not(:hover) {
            background: rgba(255, 255, 255, 0.75);
        }
    }

    /* always hide toc panel on print */
    @media only print {
        #toc_panel {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* lightbox plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* regular <img> in document when hovered */
        .lightbox_document_img:hover {
            cursor: pointer;
        }

        .body_no_scroll {
            overflow: hidden !important;
        }

        /* screen overlay */
        #lightbox_overlay {
            display: flex;
            flex-direction: column;
            position: fixed;
            left: 0;
            top: 0;
            right: 0;
            bottom: 0;
            background: rgba(0, 0, 0, 0.75);
            z-index: 3;
        }

        /* middle area containing lightbox image */
        #lightbox_image_container {
            flex-grow: 1;
            display: flex;
            justify-content: center;
            align-items: center;
            overflow: hidden;
            position: relative;
            padding: 20px;
        }

        /* bottom area containing caption */
        #lightbox_bottom_container {
            display: flex;
            justify-content: center;
            align-items: center;
            height: 100px;
            min-height: 100px;
            max-height: 100px;
            background: rgba(0, 0, 0, 0.5);
        }

        /* image number info text box */
        #lightbox_number_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            left: 2px;
            top: 0;
            z-index: 4;
        }

        /* zoom info text box */
        #lightbox_zoom_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            right: 2px;
            top: 0;
            z-index: 4;
        }

        /* copy of image caption */
        #lightbox_caption {
            box-sizing: border-box;
            display: inline-block;
            width: 100%;
            max-height: 100%;
            padding: 10px 0;
            text-align: center;
            overflow-y: auto;
            color: #ffffff;
        }

        /* navigation previous/next button */
        .lightbox_button {
            width: 100px;
            height: 100%;
            min-width: 100px;
            min-height: 100%;
            color: #ffffff;
        }

        /* navigation previous/next button when hovered */
        .lightbox_button:hover {
            background: none !important;
        }

        /* navigation button icon */
        .lightbox_button > svg {
            height: 25px;
        }

        /* figure auto-number */
        #lightbox_caption > span:first-of-type {
            font-weight: bold;
            margin-right: 5px;
        }

        /* lightbox image when hovered */
        #lightbox_img:hover {
            cursor: grab;
        }

        /* lightbox image when grabbed */
        #lightbox_img:active {
            cursor: grabbing;
        }
    }

    /* when on screen < 480px wide */
    @media only screen and (max-width: 480px) {
        /* make navigation buttons skinnier on small screens to make more room for caption text */
        .lightbox_button {
            width: 50px;
            min-width: 50px;
        }
    }

    /* always hide lightbox on print */
    @media only print {
        #lightbox_overlay {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* hypothesis (annotations) plugin */
    /* -------------------------------------------------- */

    /* hypothesis activation button */
    #hypothesis_button {
        box-sizing: border-box;
        position: fixed;
        top: 0;
        right: 0;
        width: 60px;
        height: 60px;
        background: #ffffff;
        border-radius: 0;
        border-left: solid 1px #bdbdbd;
        border-bottom: solid 1px #bdbdbd;
        box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
        z-index: 2;
    }

    /* hypothesis button svg */
    #hypothesis_button > svg {
        position: relative;
        top: -4px;
    }

    /* hypothesis annotation count */
    #hypothesis_count {
        position: absolute;
        left: 0;
        right: 0;
        bottom: 5px;
    }

    /* side panel */
    .annotator-frame {
        width: 280px !important;
    }

    /* match highlight color to rest of theme */
    .annotator-highlights-always-on .annotator-hl {
        background-color: #ffeb3b !important;
    }

    /* match focused color to rest of theme */
    .annotator-hl.annotator-hl-focused {
        background-color: #ff8a65 !important;
    }

    /* match bucket bar color to rest of theme */
    .annotator-bucket-bar {
        background: #f5f5f5 !important;
    }

    /* always hide button, toolbar, and tooltip on print */
    @media only print {
        #hypothesis_button {
            display: none;
        }

        .annotator-frame {
            display: none !important;
        }

        hypothesis-adder {
            display: none !important;
        }
    }
</style>
<!-- anchors plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds an anchor next to each of a certain type
        // of element that provides a human-readable url to that specific
        // item/position in the document (eg "manuscript.html#abstract"). It
        // also makes it such that scrolling out of view of a target removes
        // its identifier from the url.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'anchors';

        // default plugin options
        const options = {
            // which types of elements to add anchors next to, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3, [id^="fig:"], [id^="tbl:"], [id^="eq:"]',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // add anchor to each element of specified types
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements)
                addAnchor(element);

            // attach scroll listener to window
            window.addEventListener('scroll', onScroll);
        }

        // when window is scrolled
        function onScroll() {
            // if url has hash and user has scrolled out of view of hash
            // target, remove hash from url
            const tolerance = 100;
            const target = getHashTarget();
            if (target) {
                if (
                    target.getBoundingClientRect().top >
                        window.innerHeight + tolerance ||
                    target.getBoundingClientRect().bottom < 0 - tolerance
                )
                    history.pushState(null, null, ' ');
            }
        }

        // add anchor to element
        function addAnchor(element) {
            let addTo; // element to add anchor button to

            // if figure or table, modify withId and addTo to get expected
            // elements
            if (element.id.indexOf('fig:') === 0) {
                addTo = element.querySelector('figcaption');
            } else if (element.id.indexOf('tbl:') === 0) {
                addTo = element.querySelector('caption');
            } else if (element.id.indexOf('eq:') === 0) {
                addTo = element.querySelector('.eqnos-number');
            }

            addTo = addTo || element;
            const id = element.id || null;

            // do not add anchor if element doesn't have assigned id.
            // id is generated by pandoc and is assumed to be unique and
            // human-readable
            if (!id)
                return;

            // create anchor button
            const anchor = document.createElement('a');
            anchor.innerHTML = document.querySelector('.icon_link').innerHTML;
            anchor.title = 'Link to this part of the document';
            anchor.classList.add('icon_button', 'anchor');
            anchor.dataset.ignore = 'true';
            anchor.href = '#' + id;
            addTo.appendChild(anchor);
        }

        // get element that is target of link or url hash
        function getHashTarget() {
            const hash = window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');
            if (id.indexOf('tbl:') === 0)
                target = target.querySelector('table');

            return target;
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- link icon -->

<template class="icon_link">
    <!-- modified from: https://fontawesome.com/icons/link -->
    <svg width="16" height="16" viewBox="0 0 512 512">
        <path
            fill="currentColor"
            d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
        ></path>
    </svg>
</template>
<!-- accordion plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows sections of content under <h2> headings
        // to be collapsible.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'accordion';

        // default plugin options
        const options = {
            // whether to always start expanded ('false'), always start
            // collapsed ('true'), or start collapsed when screen small ('auto')
            startCollapsed: 'auto',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <h2> heading
            const headings = document.querySelectorAll('h2');
            for (const heading of headings) {
                addArrow(heading);

                // start expanded/collapsed based on option
                if (
                    options.startCollapsed === 'true' ||
                    (options.startCollapsed === 'auto' && isSmallScreen())
                )
                    collapseHeading(heading);
                else
                    expandHeading(heading);
            }

            // attach hash change listener to window
            window.addEventListener('hashchange', onHashChange);
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                goToElement(target);
        }

        // add arrow to heading
        function addArrow(heading) {
            // add arrow button
            const arrow = document.createElement('button');
            arrow.innerHTML = document.querySelector(
                '.icon_angle_down'
            ).innerHTML;
            arrow.classList.add('icon_button', 'accordion_arrow');
            heading.insertBefore(arrow, heading.firstChild);

            // attach click listener to heading and button
            heading.addEventListener('click', onHeadingClick);
            arrow.addEventListener('click', onArrowClick);
        }

        // determine if on mobile-like device with small screen
        function isSmallScreen() {
            return Math.min(window.innerWidth, window.innerHeight) < 480;
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get element that is target of hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');
            if (id.indexOf('tbl:') === 0)
                target = target.querySelector('table');

            return target;
        }

        // when <h2> heading is clicked
        function onHeadingClick(event) {
            // only collapse if <h2> itself is target of click (eg, user did
            // not click on anchor within <h2>)
            if (event.target === this)
                toggleCollapse(this);
        }

        // when arrow button is clicked
        function onArrowClick() {
            toggleCollapse(this.parentNode);
        }

        // collapse section if expanded, expand if collapsed
        function toggleCollapse(heading) {
            if (heading.dataset.collapsed === 'false')
                collapseHeading(heading);
            else
                expandHeading(heading);
        }

        // elements to exclude from collapse, such as table of contents panel,
        // hypothesis panel, etc
        const exclude = '#toc_panel, div.annotator-frame, #lightbox_overlay';

        // collapse section
        function collapseHeading(heading) {
            heading.setAttribute('data-collapsed', 'true');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'true');
        }

        // expand section
        function expandHeading(heading) {
            heading.setAttribute('data-collapsed', 'false');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'false');
        }

        // get list of elements between this <h2> and next <h2> or <h1>
        // ("children" of the <h2> section)
        function getChildren(heading) {
            return nextUntil(heading, 'h2, h1', exclude);
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get list of elements after a start element up to element matching
        // query
        function nextUntil(element, query, exclude) {
            const elements = [];
            while (element = element.nextElementSibling, element) {
                if (element.matches(query))
                    break;
                if (!element.matches(exclude))
                    elements.push(element);
            }
            return elements;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
    <!-- modified from: https://fontawesome.com/icons/angle-down -->
    <svg width="16" height="16" viewBox="0 0 448 512">
        <path
            fill="currentColor"
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        ></path>
    </svg>
</template>
<!-- tooltips plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when the user hovers or
        // focuses a link to a citation or figure, a tooltip appears with a
        // preview of the reference content, along with arrows to navigate
        // between instances of the same reference in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tooltips';

        // default plugin options
        const options = {
            // whether user must click off to close tooltip instead of just
            // un-hovering
            clickClose: 'false',
            // delay (in ms) between opening and closing tooltip
            delay: '100',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach hover and focus listeners to link
                link.addEventListener('mouseover', onLinkHover);
                link.addEventListener('mouseleave', onLinkUnhover);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('touchend', onLinkTouch);
            }

            // attach mouse, key, and resize listeners to window
            window.addEventListener('mousedown', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('keyup', onKeyUp);
            window.addEventListener('resize', onResize);
        }

        // when link is hovered
        function onLinkHover() {
            // function to open tooltip
            const delayOpenTooltip = function() {
                openTooltip(this);
            }.bind(this);

            // run open function after delay
            this.openTooltipTimer = window.setTimeout(
                delayOpenTooltip,
                options.delay
            );
        }

        // when mouse leaves link
        function onLinkUnhover() {
            // cancel opening tooltip
            window.clearTimeout(this.openTooltipTimer);

            // don't close on unhover if option specifies
            if (options.clickClose === 'true')
                return;

            // function to close tooltip
            const delayCloseTooltip = function() {
                // if tooltip open and if mouse isn't over tooltip, close
                const tooltip = document.getElementById('tooltip');
                if (tooltip && !tooltip.matches(':hover'))
                    closeTooltip();
            };

            // run close function after delay
            this.closeTooltipTimer = window.setTimeout(
                delayCloseTooltip,
                options.delay
            );
        }

        // when link is focused (tabbed to)
        function onLinkFocus(event) {
            openTooltip(this);
        }

        // when link is touched on touch screen
        function onLinkTouch(event) {
            // attempt to force hover state on first tap always, and trigger
            // regular link click (and navigation) on second tap
            if (event.target === document.activeElement)
                event.target.click();
            else {
                document.activeElement.blur();
                event.target.focus();
            }
            if (event.cancelable)
                event.preventDefault();
            event.stopPropagation();
            return false;
        }

        // when mouse is clicked anywhere in window
        function onClick(event) {
            closeTooltip();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'tooltip_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'tooltip_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeTooltip();
                    break;
            }
        }

        // when window is resized or zoomed
        function onResize() {
            closeTooltip();
        }

        // get all links of types we wish to handle
        function getLinks() {
            const queries = [];
            // exclude buttons, anchor links, toc links, etc
            const exclude =
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            queries.push('a[href^="#ref-"]' + exclude); // citation links
            queries.push('a[href^="#fig:"]' + exclude); // figure links
            const query = queries.join(', ');
            return document.querySelectorAll(query);
        }

        // get links with same target, get index of link in set, get total
        // same links
        function getSameLinks(link) {
            const sameLinks = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    sameLinks.push(otherLink);
            }

            return {
                elements: sameLinks,
                index: sameLinks.indexOf(link),
                total: sameLinks.length
            };
        }

        // open tooltip
        function openTooltip(link) {
            // delete tooltip if it exists, start fresh
            closeTooltip();

            // make tooltip element
            const tooltip = makeTooltip(link);

            // if source couldn't be found and tooltip not made, exit
            if (!tooltip)
                return;

            // make navbar elements
            const navBar = makeNavBar(link);
            if (navBar)
                tooltip.firstElementChild.appendChild(navBar);

            // attach tooltip to page
            document.body.appendChild(tooltip);

            // position tooltip
            const position = function() {
                positionTooltip(link);
            };
            position();

            // if tooltip contains images, position again after they've loaded
            const imgs = tooltip.querySelectorAll('img');
            for (const img of imgs)
                img.addEventListener('load', position);
        }

        // close (delete) tooltip
        function closeTooltip() {
            const tooltip = document.getElementById('tooltip');
            if (tooltip)
                tooltip.remove();
        }

        // make tooltip
        function makeTooltip(link) {
            // get target element that link points to
            const source = getSource(link);

            // if source can't be found, exit
            if (!source)
                return;

            // create new tooltip
            const tooltip = document.createElement('div');
            tooltip.id = 'tooltip';
            const tooltipContent = document.createElement('div');
            tooltipContent.id = 'tooltip_content';
            tooltip.appendChild(tooltipContent);

            // make copy of source node and put in tooltip
            const sourceCopy = makeCopy(source);
            tooltipContent.appendChild(sourceCopy);

            // attach mouse event listeners
            tooltip.addEventListener('click', onTooltipClick);
            tooltip.addEventListener('mousedown', onTooltipClick);
            tooltip.addEventListener('touchstart', onTooltipClick);
            tooltip.addEventListener('mouseleave', onTooltipUnhover);

            // (for interaction with lightbox plugin)
            // transfer click on tooltip copied img to original img
            const sourceImg = source.querySelector('img');
            const sourceCopyImg = sourceCopy.querySelector('img');
            if (sourceImg && sourceCopyImg) {
                const clickImg = function() {
                    sourceImg.click();
                    closeTooltip();
                };
                sourceCopyImg.addEventListener('click', clickImg);
            }

            return tooltip;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // when tooltip is clicked
        function onTooltipClick(event) {
            // when user clicks on tooltip, stop click from transferring
            // outside of tooltip (eg, click off to close tooltip, or eg click
            // off to unhighlight same refs)
            event.stopPropagation();
        }

        // when tooltip is unhovered
        function onTooltipUnhover(event) {
            if (options.clickClose === 'true')
                return;

            // make sure new mouse/touch/focus no longer over tooltip or any
            // element within it
            const tooltip = document.getElementById('tooltip');
            if (!tooltip)
                return;
            if (this.contains(event.relatedTarget))
                return;

            closeTooltip();
        }

        // make nav bar to go betwen prev/next instances of same reference
        function makeNavBar(link) {
            // find other links to the same source
            const sameLinks = getSameLinks(link);

            // don't show nav bar when singular reference
            if (sameLinks.total <= 1)
                return;

            // find prev/next links with same target
            const prevLink = getPrevLink(link, sameLinks);
            const nextLink = getNextLink(link, sameLinks);

            // create nav bar
            const navBar = document.createElement('div');
            navBar.id = 'tooltip_nav_bar';
            const text = sameLinks.index + 1 + ' of ' + sameLinks.total;

            // create nav bar prev/next buttons
            const prevButton = document.createElement('button');
            const nextButton = document.createElement('button');
            prevButton.id = 'tooltip_prev_button';
            nextButton.id = 'tooltip_next_button';
            prevButton.title =
                'Jump to the previous occurence of this item in the document [←]';
            nextButton.title =
                'Jump to the next occurence of this item in the document [→]';
            prevButton.classList.add('icon_button');
            nextButton.classList.add('icon_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;
            navBar.appendChild(prevButton);
            navBar.appendChild(document.createTextNode(text));
            navBar.appendChild(nextButton);

            // attach click listeners to buttons
            prevButton.addEventListener('click', function() {
                onPrevNextClick(link, prevLink);
            });
            nextButton.addEventListener('click', function() {
                onPrevNextClick(link, nextLink);
            });

            return navBar;
        }

        // get previous link with same target
        function getPrevLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if < 1
            let index;
            if (sameLinks.index - 1 >= 0)
                index = sameLinks.index - 1;
            else
                index = sameLinks.total - 1;
            return sameLinks.elements[index];
        }

        // get next link with same target
        function getNextLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if > total
            let index;
            if (sameLinks.index + 1 <= sameLinks.total - 1)
                index = sameLinks.index + 1;
            else
                index = 0;
            return sameLinks.elements[index];
        }

        // get element that is target of link or url hash
        function getSource(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if ref or figure, modify target to get expected element
            if (id.indexOf('ref-') === 0)
                target = target.querySelector('p');
            else if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');

            return target;
        }

        // when prev/next arrow button is clicked
        function onPrevNextClick(link, prevNextLink) {
            if (link && prevNextLink)
                goToElement(prevNextLink, window.innerHeight * 0.5);
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // determine position to place tooltip based on link position in
        // viewport and tooltip size
        function positionTooltip(link, left, top) {
            const tooltipElement = document.getElementById('tooltip');
            if (!tooltipElement)
                return;

            // get convenient vars for position/dimensions of
            // link/tooltip/page/view
            link = getRectInPage(link);
            const tooltip = getRectInPage(tooltipElement);
            const view = getRectInPage();

            // horizontal positioning
            if (left)
                // use explicit value
                left = left;
            else if (link.left + tooltip.width < view.right)
                // fit tooltip to right of link
                left = link.left;
            else if (link.right - tooltip.width > view.left)
                // fit tooltip to left of link
                left = link.right - tooltip.width;
            // center tooltip in view
            else
                left = (view.right - view.left) / 2 - tooltip.width / 2;

            // vertical positioning
            if (top)
                // use explicit value
                top = top;
            else if (link.top - tooltip.height > view.top)
                // fit tooltip above link
                top = link.top - tooltip.height;
            else if (link.bottom + tooltip.height < view.bottom)
                // fit tooltip below link
                top = link.bottom;
            else {
                // center tooltip in view
                top = view.top + view.height / 2 - tooltip.height / 2;
                // nudge off of link to left/right if possible
                if (link.right + tooltip.width < view.right)
                    left = link.right;
                else if (link.left - tooltip.width > view.left)
                    left = link.left - tooltip.width;
            }

            tooltipElement.style.left = left + 'px';
            tooltipElement.style.top = top + 'px';
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get position of element relative to page
        function getRectInPage(element) {
            const rect = getRectInView(element);
            const body = getRectInView(document.body);

            const newRect = {};
            newRect.left = rect.left - body.left;
            newRect.top = rect.top - body.top;
            newRect.right = rect.right - body.left;
            newRect.bottom = rect.bottom - body.top;
            newRect.width = rect.width;
            newRect.height = rect.height;

            return newRect;
        }

        // (for interaction with accordion plugin)
        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // (for interaction with accordion plugin)
        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- jump to first plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds a button next to each reference entry,
        // figure, and table that jumps the page to the first occurrence of a
        // link to that item in the manuscript.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'jumpToFirst';

        // default plugin options
        const options = {
            // whether to add buttons next to reference entries
            references: 'true',
            // whether to add buttons next to figures
            figures: 'true',
            // whether to add buttons next to tables
            tables: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            if (options.references !== 'false')
                makeReferenceButtons();
            if (options.figures !== 'false')
                makeFigureButtons();
            if (options.tables !== 'false')
                makeTableButtons();
        }

        // when jump button clicked
        function onButtonClick() {
            const first = getFirstOccurrence(this.dataset.id);
            if (!first)
                return;

            // update url hash so navigating "back" in history will return
            // user to jump button
            window.location.hash = this.dataset.id;
            // scroll to link
            window.setTimeout(function() {
                goToElement(first, window.innerHeight * 0.5);
            }, 0);
        }

        // get first occurence of link to item in document
        function getFirstOccurrence(id) {
            let query = 'a';
            query += '[href="#' + id + '"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelector(query);
        }

        // add button next to each reference entry
        function makeReferenceButtons() {
            const references = document.querySelectorAll('div[id^="ref-"]');
            for (const reference of references) {
                // get reference id and element to add button to
                const id = reference.id;
                const container = reference.firstElementChild;
                const first = getFirstOccurrence(id);

                // if can't find link to reference, ignore
                if (!first)
                    continue;

                // make jump button
                let button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this reference in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.innerHTML = button.outerHTML + container.innerHTML;
                button = container.firstElementChild;
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeFigureButtons() {
            const figures = document.querySelectorAll('[id^="fig:"]');
            for (const figure of figures) {
                // get figure id and element to add button to
                const id = figure.id;
                const container = figure.querySelector('figcaption') || figure;
                const first = getFirstOccurrence(id);

                // if can't find link to figure, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this figure in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeTableButtons() {
            const tables = document.querySelectorAll('[id^="tbl:"]');
            for (const table of tables) {
                // get ref id and element to add button to
                const id = table.id;
                const container = table.querySelector('caption') || table;
                const first = getFirstOccurrence(id);

                // if can't find link to table, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this table in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
    <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
    <svg width="16" height="16" viewBox="0 0 320 512">
        <path
            fill="currentColor"
            d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
        ></path>
    </svg>
</template>
<!-- link highlight plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user hovers or
        // focuses a link, other links that have the same target will be
        // highlighted. It also makes it such that when clicking a link, the
        // target of the link (eg reference, figure, table) is briefly
        // highlighted.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'linkHighlight';

        // default plugin options
        const options = {
            // whether to also highlight links that go to external urls
            externalLinks: 'false',
            // whether user must click off to unhighlight instead of just
            // un-hovering
            clickUnhighlight: 'false',
            // whether to also highlight links that are unique
            highlightUnique: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach mouse and focus listeners to link
                link.addEventListener('mouseenter', onLinkFocus);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('mouseleave', onLinkUnhover);
            }

            // attach click and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('hashchange', onHashChange);

            // run hash change on window load in case user has navigated
            // directly to hash
            onHashChange();
        }

        // when link is focused (tabbed to) or hovered
        function onLinkFocus() {
            highlight(this);
        }

        // when link is unhovered
        function onLinkUnhover() {
            if (options.clickUnhighlight !== 'true')
                unhighlightAll();
        }

        // when the mouse is clicked anywhere in window
        function onClick(event) {
            unhighlightAll();
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                glowElement(target);
        }

        // get element that is target of link or url hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            return target;
        }

        // start glow sequence on an element
        function glowElement(element) {
            const startGlow = function() {
                onGlowEnd();
                element.dataset.glow = 'true';
                element.addEventListener('animationend', onGlowEnd);
            };
            const onGlowEnd = function() {
                element.removeAttribute('data-glow');
                element.removeEventListener('animationend', onGlowEnd);
            };
            startGlow();
        }

        // highlight link and all others with same target
        function highlight(link) {
            // force unhighlight all to start fresh
            unhighlightAll();

            // get links with same target
            if (!link)
                return;
            const sameLinks = getSameLinks(link);

            // if link unique and option is off, exit and don't highlight
            if (sameLinks.length <= 1 && options.highlightUnique !== 'true')
                return;

            // highlight all same links, and "select" (special highlight) this
            // one
            for (const sameLink of sameLinks) {
                if (sameLink === link)
                    sameLink.setAttribute('data-selected', 'true');
                else
                    sameLink.setAttribute('data-highlighted', 'true');
            }
        }

        // unhighlight all links
        function unhighlightAll() {
            const links = getLinks();
            for (const link of links) {
                link.setAttribute('data-selected', 'false');
                link.setAttribute('data-highlighted', 'false');
            }
        }

        // get links with same target
        function getSameLinks(link) {
            const results = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    results.push(otherLink);
            }
            return results;
        }

        // get all links of types we wish to handle
        function getLinks() {
            let query = 'a';
            if (options.externalLinks !== 'true')
                query += '[href^="#"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelectorAll(query);
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- table of contents plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin provides a "table of contents" (toc) panel on
        // the side of the document that allows the user to conveniently
        // navigate between sections of the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tableOfContents';

        // default plugin options
        const options = {
            // which types of elements to add links for, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3',
            // whether default behavior is to be closed ('false'), open
            // ('true'), or only open when screen wide enough to fit panel
            // ('auto'). note: still always starts closed when page loads.
            open: 'auto',
            // if list item is more than this many characters, text will be
            // truncated
            charLimit: '50',
            // whether or not to show bullets next to each toc item
            bullets: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // make toc panel and populate with entries (links to document
            // sections)
            const panel = makePanel();
            if (!panel)
                return;
            makeEntries(panel);
            document.body.insertBefore(panel, document.body.firstChild);

            closePanel();

            // attach click, scroll, and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('scroll', onScroll);
            window.addEventListener('hashchange', onScroll);
            window.addEventListener('keyup', onKeyUp);
            onScroll();

            // add class to push document body down out of way of toc button
            document.body.classList.add('toc_body_nudge');
        }

        // determine if screen wide enough to fit toc panel
        function isSmallScreen() {
            // in default theme:
            // 816px = 8.5in = width of "page" (<body>) element
            // 260px = min width of toc panel (*2 for both sides of <body>)
            return window.innerWidth < 816 + 260 * 2;
        }

        // open/close panel based on option and screen size
        function openOrClosePanel() {
            if (
                options.open === 'true' ||
                (options.open === 'auto' && !isSmallScreen())
            )
                openPanel();
            else
                closePanel();
        }

        // when mouse is clicked anywhere in window
        function onClick() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                openOrClosePanel();
        }

        // when window is scrolled or hash changed
        function onScroll() {
            highlightViewed();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            // close on esc
            if (event.key === 'Escape')
                closePanel();
        }

        // find entry of currently viewed document section in toc and highlight
        function highlightViewed() {
            const firstId = getFirstInView(options.typesQuery);

            // get toc entries (links), unhighlight all, then highlight viewed
            const list = document.getElementById('toc_list');
            if (!firstId || !list)
                return;
            const links = list.querySelectorAll('a');
            for (const link of links)
                link.dataset.viewing = 'false';
            const link = list.querySelector('a[href="#' + firstId + '"]');
            if (!link)
                return;
            link.dataset.viewing = 'true';
        }

        // get first or previous toc listed element in top half of view
        function getFirstInView(query) {
            // get all elements matching query and with id
            const elements = document.querySelectorAll(query);
            const elementsWithIds = [];
            for (const element of elements) {
                if (element.id)
                    elementsWithIds.push(element);
            }


            // get first or previous element in top half of view
            for (let i = 0; i < elementsWithIds.length; i++) {
                const element = elementsWithIds[i];
                const prevElement = elementsWithIds[Math.max(0, i - 1)];
                if (element.getBoundingClientRect().top >= 0) {
                    if (
                        element.getBoundingClientRect().top <
                        window.innerHeight / 2
                    )
                        return element.id;
                    else
                        return prevElement.id;
                }
            }
        }

        // make panel
        function makePanel() {
            // create panel
            const panel = document.createElement('div');
            panel.id = 'toc_panel';
            if (options.bullets === 'true')
                panel.dataset.bullets = 'true';

            // create header
            const header = document.createElement('div');
            header.id = 'toc_header';

            // create toc button
            const button = document.createElement('button');
            button.id = 'toc_button';
            button.innerHTML = document.querySelector('.icon_th_list').innerHTML;
            button.title = 'Table of Contents';
            button.classList.add('icon_button');

            // create header text
            const text = document.createElement('h3');
            text.innerHTML = 'View Table of Contents';

            // create container for toc list
            const list = document.createElement('div');
            list.id = 'toc_list';

            // attach click listeners
            panel.addEventListener('click', onPanelClick);
            header.addEventListener('click', onHeaderClick);
            button.addEventListener('click', onButtonClick);

            // attach elements
            header.appendChild(button);
            header.appendChild(text);
            panel.appendChild(header);
            panel.appendChild(list);

            return panel;
        }

        // create toc entries (links) to each element of the specified types
        function makeEntries(panel) {
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements) {
                // do not add link if element doesn't have assigned id
                if (!element.id)
                    continue;

                // create link/list item
                const link = document.createElement('a');
                link.classList.add('toc_link');
                switch (element.tagName.toLowerCase()) {
                    case 'h1':
                        link.dataset.level = '1';
                        break;
                    case 'h2':
                        link.dataset.level = '2';
                        break;
                    case 'h3':
                        link.dataset.level = '3';
                        break;
                    case 'h4':
                        link.dataset.level = '4';
                        break;
                }
                link.title = element.innerText;
                let text = element.innerText;
                if (text.length > options.charLimit)
                    text = text.slice(0, options.charLimit) + '...';
                link.innerHTML = text;
                link.href = '#' + element.id;
                link.addEventListener('click', onLinkClick);

                // attach link
                panel.querySelector('#toc_list').appendChild(link);
            }
        }

        // when panel is clicked
        function onPanelClick(event) {
            // stop click from propagating to window/document and closing panel
            event.stopPropagation();
        }

        // when header itself is clicked
        function onHeaderClick(event) {
            togglePanel();
        }

        // when button is clicked
        function onButtonClick(event) {
            togglePanel();
            // stop header underneath button from also being clicked
            event.stopPropagation();
        }

        // when link is clicked
        function onLinkClick() {
            openOrClosePanel();
        }

        // open panel if closed, close if opened
        function togglePanel() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                closePanel();
            else
                openPanel();
        }

        // open panel
        function openPanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'true';
        }

        // close panel
        function closePanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'false';
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- th list icon -->

<template class="icon_th_list">
    <!-- modified from: https://fontawesome.com/icons/th-list -->
    <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
        <path
            fill="currentColor"
            d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
            tabindex="-1"
        ></path>
    </svg>
</template>
<!-- lightbox plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user clicks on an
        // image, the image fills the screen and the user can pan/drag/zoom
        // the image and navigate between other images in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'lightbox';

        // default plugin options
        const options = {
            // list of possible zoom/scale factors
            zoomSteps:
                '0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1,' +
                '1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8',
            // whether to fit image to view ('fit'), display at 100% and shrink
            // if necessary ('shrink'), or always display at 100% ('100')
            defaultZoom: 'fit',
            // whether to zoom in/out toward center of view ('true') or mouse
            // ('false')
            centerZoom: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <img> element
            const imgs = document.querySelectorAll('figure > img');
            let count = 1;
            for (const img of imgs) {
                img.classList.add('lightbox_document_img');
                img.dataset.number = count;
                img.dataset.total = imgs.length;
                img.addEventListener('click', openLightbox);
                count++;
            }

            // attach mouse and key listeners to window
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('keyup', onKeyUp);
        }

        // when mouse is moved anywhere in window
        function onWindowMouseMove(event) {
            window.mouseX = event.clientX;
            window.mouseY = event.clientY;
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'lightbox_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'lightbox_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeLightbox();
                    break;
            }
        }

        // open lightbox
        function openLightbox() {
            const lightbox = makeLightbox(this);
            if (!lightbox)
                return;

            blurBody(lightbox);
            document.body.appendChild(lightbox);
        }

        // make lightbox
        function makeLightbox(img) {
            // delete lightbox if it exists, start fresh
            closeLightbox();

            // create screen overlay containing lightbox
            const overlay = document.createElement('div');
            overlay.id = 'lightbox_overlay';

            // create image info boxes
            const numberInfo = document.createElement('div');
            const zoomInfo = document.createElement('div');
            numberInfo.id = 'lightbox_number_info';
            zoomInfo.id = 'lightbox_zoom_info';

            // create container for image
            const imageContainer = document.createElement('div');
            imageContainer.id = 'lightbox_image_container';
            const lightboxImg = makeLightboxImg(
                img,
                imageContainer,
                numberInfo,
                zoomInfo
            );
            imageContainer.appendChild(lightboxImg);

            // create bottom container for caption and navigation buttons
            const bottomContainer = document.createElement('div');
            bottomContainer.id = 'lightbox_bottom_container';
            const caption = makeCaption(img);
            const prevButton = makePrevButton(img);
            const nextButton = makeNextButton(img);
            bottomContainer.appendChild(prevButton);
            bottomContainer.appendChild(caption);
            bottomContainer.appendChild(nextButton);

            // attach top middle and bottom to overlay
            overlay.appendChild(numberInfo);
            overlay.appendChild(zoomInfo);
            overlay.appendChild(imageContainer);
            overlay.appendChild(bottomContainer);

            return overlay;
        }

        // make <img> object that is intuitively draggable and zoomable
        function makeLightboxImg(
            sourceImg,
            container,
            numberInfoBox,
            zoomInfoBox
        ) {
            // create copy of source <img>
            const img = sourceImg.cloneNode(true);
            img.classList.remove('lightbox_document_img');
            img.removeAttribute('id');
            img.removeAttribute('width');
            img.removeAttribute('height');
            img.style.position = 'unset';
            img.style.margin = '0';
            img.style.padding = '0';
            img.style.width = '';
            img.style.height = '';
            img.style.minWidth = '';
            img.style.minHeight = '';
            img.style.maxWidth = '';
            img.style.maxHeight = '';
            img.id = 'lightbox_img';

            // build sorted list of unique zoomSteps, always including a 100%
            let zoomSteps = [];
            const optionsZooms = options.zoomSteps.split(/[^0-9.]/);
            for (const optionZoom of optionsZooms) {
                const newZoom = parseFloat(optionZoom);
                if (newZoom && !zoomSteps.includes(newZoom))
                    zoomSteps.push(newZoom);
            }
            if (!zoomSteps.includes(1))
                zoomSteps.push(1);
            zoomSteps = zoomSteps.sort(function sortNumber(a, b) {
                return a - b;
            });

            // <img> object property variables
            let zoom = 1;
            let translateX = 0;
            let translateY = 0;
            let clickMouseX = undefined;
            let clickMouseY = undefined;
            let clickTranslateX = undefined;
            let clickTranslateY = undefined;

            updateNumberInfo();

            // update image numbers displayed in info box
            function updateNumberInfo() {
                numberInfoBox.innerHTML =
                    sourceImg.dataset.number + ' of ' + sourceImg.dataset.total;
            }

            // update zoom displayed in info box
            function updateZoomInfo() {
                let zoomInfo = zoom * 100;
                if (!Number.isInteger(zoomInfo))
                    zoomInfo = zoomInfo.toFixed(2);
                zoomInfoBox.innerHTML = zoomInfo + '%';
            }

            // move to closest zoom step above current zoom
            const zoomIn = function() {
                for (const zoomStep of zoomSteps) {
                    if (zoomStep > zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                updateTransform();
            };

            // move to closest zoom step above current zoom
            const zoomOut = function() {
                zoomSteps.reverse();
                for (const zoomStep of zoomSteps) {
                    if (zoomStep < zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                zoomSteps.reverse();

                updateTransform();
            };

            // update display of <img> based on scale/translate properties
            const updateTransform = function() {
                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                // get new width/height after scale
                const rect = img.getBoundingClientRect();
                // limit translate
                translateX = Math.max(translateX, -rect.width / 2);
                translateX = Math.min(translateX, rect.width / 2);
                translateY = Math.max(translateY, -rect.height / 2);
                translateY = Math.min(translateY, rect.height / 2);

                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                updateZoomInfo();
            };

            // fit <img> to container
            const fit = function() {
                // no x/y offset, 100% zoom by default
                translateX = 0;
                translateY = 0;
                zoom = 1;

                // widths of <img> and container
                const imgWidth = img.naturalWidth;
                const imgHeight = img.naturalHeight;
                const containerWidth = parseFloat(
                    window.getComputedStyle(container).width
                );
                const containerHeight = parseFloat(
                    window.getComputedStyle(container).height
                );

                // how much zooming is needed to fit <img> to container
                const xRatio = imgWidth / containerWidth;
                const yRatio = imgHeight / containerHeight;
                const maxRatio = Math.max(xRatio, yRatio);
                const newZoom = 1 / maxRatio;

                // fit <img> to container according to option
                if (options.defaultZoom === 'shrink') {
                    if (maxRatio > 1)
                        zoom = newZoom;
                } else if (options.defaultZoom === 'fit')
                    zoom = newZoom;

                updateTransform();
            };

            // when mouse wheel is rolled anywhere in container
            const onContainerWheel = function(event) {
                if (!event)
                    return;

                // let ctrl + mouse wheel to zoom behave as normal
                if (event.ctrlKey)
                    return;

                // prevent normal scroll behavior
                event.preventDefault();
                event.stopPropagation();

                // point around which to scale img
                const viewRect = container.getBoundingClientRect();
                const viewX = (viewRect.left + viewRect.right) / 2;
                const viewY = (viewRect.top + viewRect.bottom) / 2;
                const originX = options.centerZoom === 'true' ? viewX : mouseX;
                const originY = options.centerZoom === 'true' ? viewY : mouseY;

                // get point on image under origin
                const oldRect = img.getBoundingClientRect();
                const oldPercentX = (originX - oldRect.left) / oldRect.width;
                const oldPercentY = (originY - oldRect.top) / oldRect.height;

                // increment/decrement zoom
                if (event.deltaY < 0)
                    zoomIn();
                if (event.deltaY > 0)
                    zoomOut();

                // get offset between previous image point and origin
                const newRect = img.getBoundingClientRect();
                const offsetX =
                    originX - (newRect.left + newRect.width * oldPercentX);
                const offsetY =
                    originY - (newRect.top + newRect.height * oldPercentY);

                // translate image to keep image point under origin
                translateX += offsetX;
                translateY += offsetY;

                // perform translate
                updateTransform();
            };

            // when container is clicked
            function onContainerClick(event) {
                // if container itself is target of click, and not other
                // element above it
                if (event.target === this)
                    closeLightbox();
            }

            // when mouse button is pressed on image
            const onImageMouseDown = function(event) {
                // store original mouse position relative to image
                clickMouseX = window.mouseX;
                clickMouseY = window.mouseY;
                clickTranslateX = translateX;
                clickTranslateY = translateY;
                event.stopPropagation();
                event.preventDefault();
            };

            // when mouse button is released anywhere in window
            const onWindowMouseUp = function(event) {
                // reset original mouse position
                clickMouseX = undefined;
                clickMouseY = undefined;
                clickTranslateX = undefined;
                clickTranslateY = undefined;

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mouseup', onWindowMouseUp);
            };

            // when mouse is moved anywhere in window
            const onWindowMouseMove = function(event) {
                if (
                    clickMouseX === undefined ||
                    clickMouseY === undefined ||
                    clickTranslateX === undefined ||
                    clickTranslateY === undefined
                )
                    return;

                // offset image based on original and current mouse position
                translateX = clickTranslateX + window.mouseX - clickMouseX;
                translateY = clickTranslateY + window.mouseY - clickMouseY;
                updateTransform();
                event.preventDefault();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mousemove', onWindowMouseMove);
            };

            // when window is resized
            const onWindowResize = function(event) {
                fit();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('resize', onWindowResize);
            };

            // attach the necessary event listeners
            img.addEventListener('dblclick', fit);
            img.addEventListener('mousedown', onImageMouseDown);
            container.addEventListener('wheel', onContainerWheel);
            container.addEventListener('mousedown', onContainerClick);
            container.addEventListener('touchstart', onContainerClick);
            window.addEventListener('mouseup', onWindowMouseUp);
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('resize', onWindowResize);

            // run fit() after lightbox atttached to document and <img> Loaded
            // so needed container and img dimensions available
            img.addEventListener('load', fit);

            return img;
        }

        // make caption
        function makeCaption(img) {
            const caption = document.createElement('div');
            caption.id = 'lightbox_caption';
            const captionSource = img.nextElementSibling;
            if (captionSource.tagName.toLowerCase() === 'figcaption') {
                const captionCopy = makeCopy(captionSource);
                caption.innerHTML = captionCopy.innerHTML;
            }

            caption.addEventListener('touchstart', function(event) {
                event.stopPropagation();
            });

            return caption;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // make button to jump to previous image in document
        function makePrevButton(img) {
            const prevButton = document.createElement('button');
            prevButton.id = 'lightbox_prev_button';
            prevButton.title = 'Jump to the previous image in the document [←]';
            prevButton.classList.add('icon_button', 'lightbox_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;

            // attach click listeners to button
            prevButton.addEventListener('click', function() {
                getPrevImg(img).click();
            });

            return prevButton;
        }

        // make button to jump to next image in document
        function makeNextButton(img) {
            const nextButton = document.createElement('button');
            nextButton.id = 'lightbox_next_button';
            nextButton.title = 'Jump to the next image in the document [→]';
            nextButton.classList.add('icon_button', 'lightbox_button');
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;

            // attach click listeners to button
            nextButton.addEventListener('click', function() {
                getNextImg(img).click();
            });

            return nextButton;
        }

        // get previous image in document
        function getPrevImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if < 1
            if (index - 1 >= 0)
                index--;
            else
                index = imgs.length - 1;
            return imgs[index];
        }

        // get next image in document
        function getNextImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if > total
            if (index + 1 <= imgs.length - 1)
                index++;
            else
                index = 0;
            return imgs[index];
        }

        // close lightbox
        function closeLightbox() {
            focusBody();

            const lightbox = document.getElementById('lightbox_overlay');
            if (lightbox)
                lightbox.remove();
        }

        // make all elements behind lightbox non-focusable
        function blurBody(overlay) {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.tabIndex = -1;
            document.body.classList.add('body_no_scroll');
        }

        // make all elements focusable again
        function focusBody() {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.removeAttribute('tabIndex');
            document.body.classList.remove('body_no_scroll');
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- attributes plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows arbitrary HTML attributes to be attached
        // to (almost) any element. Place an HTML comment inside or next to the
        // desired element in the format <!-- $attribute="value" -->

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'attributes';

        // default plugin options
        const options = {
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // get list of comments in document
            const comments = findComments();

            for(const comment of comments)
                if (comment.parentElement)
                    addAttributes(
                        comment.parentElement,
                        comment.nodeValue.trim()
                    );
        }

        // add html attributes to specified element based on string of 
        // html attributes and values
        function addAttributes(element, text) {
            // regex's for finding attribute/value pairs in the format of
            // attribute="value" or attribute='value
            const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
            const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

            // loop through attribute/value pairs
            let match;
            while(match = text.match(regex2) || text.match(regex1)) {
                // get attribute and value from regex capture groups
                let attribute = match[1];
                let value = match[2];

                // remove from string
                text = text.substring(match.index + match[0].length);

                if (!attribute || !value)
                    break;

                // set attribute of parent element
                try {
                    element.setAttribute(attribute, value);
                } catch(error) {
                    console.log(error);
                }

                // special case for colspan
                if (attribute === 'colspan')
                    removeTableCells(element, value);
            }
        }

        // get list of comment elements in document
        function findComments() {
            const comments = [];

            // iterate over comment nodes in document
            function acceptNode(node) {
                return NodeFilter.FILTER_ACCEPT;
            }
            const iterator = document.createNodeIterator(
                document.body,
                NodeFilter.SHOW_COMMENT,
                acceptNode
            );
            let node;
            while(node = iterator.nextNode())
                comments.push(node);

            return comments;
        }

        // remove certain number of cells after specified cell
        function removeTableCells(cell, number) {
            number = parseInt(number);
            if (!number)
                return;

            // remove elements
            for(; number > 1; number--) {
                if (cell.nextElementSibling)
                    cell.nextElementSibling.remove();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- math plugin configuration -->

<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        "CommonHTML": { linebreaks: { automatic: true } },
        "HTML-CSS": { linebreaks: { automatic: true } },
        "SVG": { linebreaks: { automatic: true } },
        "fast-preview": { disabled: true }
    });
</script>

<!-- math plugin -->

<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML'>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'MathJax' allows the proper rendering of
    // math/equations written in LaTeX.

    // https://www.mathjax.org/
</script>
<!-- annotations plugin -->

<script>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'Hypothesis' allows public annotation of the
    // manuscript.

    // https://web.hypothes.is/

    // plugin configuration
    window.hypothesisConfig = function() {
        return {
            branding: {
                accentColor: '#2196f3',
                appBackgroundColor: '#f8f8f8',
                ctaBackgroundColor: '#f8f8f8',
                ctaTextColor: '#000000',
                selectionFontFamily: 'Open Sans, Helvetica, sans serif',
                annotationFontFamily: 'Open Sans, Helvetica, sans serif'
            }
        };
    };

    // hypothesis client script
    const embed = 'https://hypothes.is/embed.js';
    // hypothesis annotation count query url
    const query = 'https://api.hypothes.is/api/search?limit=0&url='

    
    // start script
    function start() {
        const button = makeButton();
        document.body.insertBefore(button, document.body.firstChild);
        insertCount(button);
    }

    // make button
    function makeButton() {
        // create button
        const button = document.createElement('button');
        button.id = 'hypothesis_button';
        button.innerHTML = document.querySelector('.icon_hypothesis').innerHTML;
        button.title = 'Hypothesis annotations';
        button.classList.add('icon_button');

        function onClick(event) {
            onButtonClick(event, button);
        }

        // attach click listeners
        button.addEventListener('click', onClick);

        return button;
    }

    // insert annotations count
    async function insertCount(button) {
        // get annotation count from Hypothesis based on url
        let count = '-';
        try {
            const canonical = document.querySelector('link[rel="canonical"]');
            const location = window.location;
            const url = encodeURIComponent((canonical || location).href);
            const response = await fetch(query + url);
            const json = await response.json();
            count = json.total || '-';
        } catch(error) {
            console.log(error);
        }
        
        // put count into button
        const counter = document.createElement('span');
        counter.id = 'hypothesis_count';
        counter.innerHTML = count;
        button.title = 'View ' + count + ' Hypothesis annotations';
        button.append(counter);
    }

    // when button is clicked
    function onButtonClick(event, button) {
        const script = document.createElement('script');
        script.src = embed;
        document.body.append(script);
        button.remove();
    }

    window.addEventListener('load', start);
</script>

<!-- hypothesis icon -->

<template class="icon_hypothesis">
    <!-- modified from: https://simpleicons.org/icons/hypothesis.svg / https://git.io/Jf1VB -->
    <svg width="16" height="16" viewBox="0 0 24 24" tabindex="-1">
        <path
            fill="currentColor"
            d="M3.43 0C2.5 0 1.72 .768 1.72 1.72V18.86C1.72 19.8 2.5 20.57 3.43 20.57H9.38L12 24L14.62 20.57H20.57C21.5 20.57 22.29 19.8 22.29 18.86V1.72C22.29 .77 21.5 0 20.57 0H3.43M5.14 3.43H7.72V9.43S8.58 7.72 10.28 7.72C12 7.72 13.74 8.57 13.74 11.24V17.14H11.16V12C11.16 10.61 10.28 10.07 9.43 10.29C8.57 10.5 7.72 11.41 7.72 13.29V17.14H5.14V3.43M18 13.72C18.95 13.72 19.72 14.5 19.72 15.42A1.71 1.71 0 0 1 18 17.13A1.71 1.71 0 0 1 16.29 15.42C16.29 14.5 17.05 13.71 18 13.71Z"
            tabindex="-1"
        ></path>
    </svg>
</template>
<!-- analytics plugin -->

<!-- copy and paste code from Google Analytics or similar service here -->
</body>
</html>
